

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>How do I … &#8212; sgkit  documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/typehints.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/docsearch.sbt.css" />
    <link rel="stylesheet" type="text/css" href="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.css" />
  
  <!-- So that users can add custom icons -->
  <script src="_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'how_do_i';</script>
    <script>
        DOCUMENTATION_OPTIONS.theme_version = '0.16.1';
        DOCUMENTATION_OPTIONS.theme_switcher_json_url = 'https://sgkit-dev.github.io/sgkit/latest/_static/switcher.json';
        DOCUMENTATION_OPTIONS.theme_switcher_version_match = 'latest';
        DOCUMENTATION_OPTIONS.show_version_warning_banner =
            false;
        </script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.js"></script>
    <script defer="defer" src="_static/docsearch.sbt.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Contributing to sgkit" href="contributing.html" />
    <link rel="prev" title="sgkit.variables.window_stop_spec" href="generated/sgkit.variables.window_stop_spec.html" /> 
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="latest" />   
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/sgkit_trnsprnt.png" class="logo__image only-light" alt="sgkit  documentation - Home"/>
    <img src="_static/sgkit_blue_trnsprnt.png" class="logo__image only-dark pst-js-only" alt="sgkit  documentation - Home"/>
  
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="getting_started.html">
    Getting Started
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="user_guide.html">
    User Guide
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="examples/index.html">
    Examples
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="api.html">
    API reference
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="#">
    How do I …
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button"
                data-bs-toggle="dropdown" aria-expanded="false"
                aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="contributing.html">
    Contributing to sgkit
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="about.html">
    About
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="news.html">
    News
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="changelog.html">
    Changelog
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">
<div class="version-switcher__container dropdown pst-js-only">
  <button id="pst-version-switcher-button-2"
    type="button"
    class="version-switcher__button btn btn-sm dropdown-toggle"
    data-bs-toggle="dropdown"
    aria-haspopup="listbox"
    aria-controls="pst-version-switcher-list-2"
    aria-label="Version switcher list"
  >
    Choose version  <!-- this text may get changed later by javascript -->
    <span class="caret"></span>
  </button>
  <div id="pst-version-switcher-list-2"
    class="version-switcher__menu dropdown-menu list-group-flush py-0"
    role="listbox" aria-labelledby="pst-version-switcher-button-2">
    <!-- dropdown will be populated by javascript on page load -->
  </div>
</div></div>
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/sgkit-dev/sgkit" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="getting_started.html">
    Getting Started
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="user_guide.html">
    User Guide
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="examples/index.html">
    Examples
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="api.html">
    API reference
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="#">
    How do I …
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="contributing.html">
    Contributing to sgkit
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="about.html">
    About
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="news.html">
    News
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="changelog.html">
    Changelog
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">
<div class="version-switcher__container dropdown pst-js-only">
  <button id="pst-version-switcher-button-3"
    type="button"
    class="version-switcher__button btn btn-sm dropdown-toggle"
    data-bs-toggle="dropdown"
    aria-haspopup="listbox"
    aria-controls="pst-version-switcher-list-3"
    aria-label="Version switcher list"
  >
    Choose version  <!-- this text may get changed later by javascript -->
    <span class="caret"></span>
  </button>
  <div id="pst-version-switcher-list-3"
    class="version-switcher__menu dropdown-menu list-group-flush py-0"
    role="listbox" aria-labelledby="pst-version-switcher-button-3">
    <!-- dropdown will be populated by javascript on page load -->
  </div>
</div></div>
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/sgkit-dev/sgkit" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"></div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">How do I …</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                   <section id="how-do-i">
<span id="id1"></span><h1>How do I …<a class="headerlink" href="#how-do-i" title="Permalink to this heading">#</a></h1>
<nav class="contents local" id="contents">
<ul class="simple">
<li><p><a class="reference internal" href="#create-a-test-dataset" id="id2">Create a test dataset?</a></p></li>
<li><p><a class="reference internal" href="#look-at-the-dataset-summary" id="id3">Look at the dataset summary?</a></p></li>
<li><p><a class="reference internal" href="#get-the-values-for-a-variable-in-a-dataset" id="id4">Get the values for a variable in a dataset?</a></p></li>
<li><p><a class="reference internal" href="#find-the-definition-for-a-variable-in-a-dataset" id="id5">Find the definition for a variable in a dataset?</a></p></li>
<li><p><a class="reference internal" href="#look-at-the-genotypes" id="id6">Look at the genotypes?</a></p></li>
<li><p><a class="reference internal" href="#subset-the-variables" id="id7">Subset the variables?</a></p></li>
<li><p><a class="reference internal" href="#subset-to-a-genomic-range" id="id8">Subset to a genomic range?</a></p></li>
<li><p><a class="reference internal" href="#get-the-list-of-samples" id="id9">Get the list of samples?</a></p></li>
<li><p><a class="reference internal" href="#subset-the-samples" id="id10">Subset the samples?</a></p></li>
<li><p><a class="reference internal" href="#define-a-new-variable-based-on-others" id="id11">Define a new variable based on others?</a></p></li>
<li><p><a class="reference internal" href="#get-summary-stats" id="id12">Get summary stats?</a></p></li>
<li><p><a class="reference internal" href="#filter-variants" id="id13">Filter variants?</a></p></li>
<li><p><a class="reference internal" href="#find-which-new-variables-were-added-by-a-method" id="id14">Find which new variables were added by a method?</a></p></li>
<li><p><a class="reference internal" href="#save-results-to-a-zarr-file" id="id15">Save results to a Zarr file?</a></p></li>
<li><p><a class="reference internal" href="#load-a-dataset-from-zarr" id="id16">Load a dataset from Zarr?</a></p></li>
</ul>
</nav>
<section id="create-a-test-dataset">
<h2><a class="toc-backref" href="#id2" role="doc-backlink">Create a test dataset?</a><a class="headerlink" href="#create-a-test-dataset" title="Permalink to this heading">#</a></h2>
<p>Call <a class="reference internal" href="generated/sgkit.simulate_genotype_call_dataset.html#sgkit.simulate_genotype_call_dataset" title="sgkit.simulate_genotype_call_dataset"><code class="xref py py-func docutils literal notranslate"><span class="pre">simulate_genotype_call_dataset()</span></code></a> to create a test <a class="reference external" href="https://docs.xarray.dev/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v2025.1.2.dev0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">xarray.Dataset</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [1]: </span><span class="kn">import</span><span class="w"> </span><span class="nn">sgkit</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sg</span>

<span class="gp">In [2]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">simulate_genotype_call_dataset</span><span class="p">(</span><span class="n">n_variant</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">n_sample</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">n_contig</span><span class="o">=</span><span class="mi">23</span><span class="p">,</span> <span class="n">missing_pct</span><span class="o">=</span><span class="mf">.1</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="look-at-the-dataset-summary">
<h2><a class="toc-backref" href="#id3" role="doc-backlink">Look at the dataset summary?</a><a class="headerlink" href="#look-at-the-dataset-summary" title="Permalink to this heading">#</a></h2>
<p>Print using the <a class="reference external" href="https://docs.xarray.dev/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v2025.1.2.dev0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">xarray.Dataset</span></code></a> <code class="docutils literal notranslate"><span class="pre">repr</span></code>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [3]: </span><span class="n">ds</span>
<span class="gh">Out[3]: </span>
<span class="go">&lt;xarray.Dataset&gt; Size: 23kB</span>
<span class="go">Dimensions:             (contigs: 23, variants: 100, alleles: 2, samples: 50,</span>
<span class="go">                         ploidy: 2)</span>
<span class="go">Dimensions without coordinates: contigs, variants, alleles, samples, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    contig_id           (contigs) &lt;U2 184B &#39;0&#39; &#39;1&#39; &#39;2&#39; &#39;3&#39; ... &#39;20&#39; &#39;21&#39; &#39;22&#39;</span>
<span class="go">    variant_contig      (variants) int64 800B 0 0 0 0 0 1 ... 21 21 22 22 22 22</span>
<span class="go">    variant_position    (variants) int64 800B 0 1 2 3 4 0 1 2 ... 1 2 3 0 1 2 3</span>
<span class="go">    variant_allele      (variants, alleles) |S1 200B b&#39;T&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    sample_id           (samples) &lt;U3 600B &#39;S0&#39; &#39;S1&#39; &#39;S2&#39; ... &#39;S47&#39; &#39;S48&#39; &#39;S49&#39;</span>
<span class="go">    call_genotype       (variants, samples, ploidy) int8 10kB 0 0 1 0 ... 1 0 0</span>
<span class="go">    call_genotype_mask  (variants, samples, ploidy) bool 10kB False ... False</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [&#39;0&#39;, &#39;1&#39;, &#39;2&#39;, &#39;3&#39;, &#39;4&#39;, &#39;5&#39;, &#39;6&#39;, &#39;7&#39;, &#39;8&#39;, &#39;9&#39;, &#39;10&#39;, &#39;11&#39;, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
</section>
<section id="get-the-values-for-a-variable-in-a-dataset">
<h2><a class="toc-backref" href="#id4" role="doc-backlink">Get the values for a variable in a dataset?</a><a class="headerlink" href="#get-the-values-for-a-variable-in-a-dataset" title="Permalink to this heading">#</a></h2>
<p>Call <a class="reference external" href="https://docs.xarray.dev/en/stable/generated/xarray.Variable.values.html#xarray.Variable.values" title="(in xarray v2025.1.2.dev0)"><code class="xref py py-attr docutils literal notranslate"><span class="pre">xarray.Variable.values</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [4]: </span><span class="n">ds</span><span class="o">.</span><span class="n">variant_contig</span><span class="o">.</span><span class="n">values</span>
<span class="gh">Out[4]: </span>
<span class="go">array([ 0,  0,  0,  0,  0,  1,  1,  1,  1,  1,  2,  2,  2,  2,  2,  3,  3,</span>
<span class="go">        3,  3,  3,  4,  4,  4,  4,  4,  5,  5,  5,  5,  5,  6,  6,  6,  6,</span>
<span class="go">        6,  7,  7,  7,  7,  7,  8,  8,  8,  8,  9,  9,  9,  9, 10, 10, 10,</span>
<span class="go">       10, 11, 11, 11, 11, 12, 12, 12, 12, 13, 13, 13, 13, 14, 14, 14, 14,</span>
<span class="go">       15, 15, 15, 15, 16, 16, 16, 16, 17, 17, 17, 17, 18, 18, 18, 18, 19,</span>
<span class="go">       19, 19, 19, 20, 20, 20, 20, 21, 21, 21, 21, 22, 22, 22, 22])</span>

<span class="gp">In [5]: </span><span class="n">ds</span><span class="p">[</span><span class="s2">&quot;variant_contig&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span> <span class="c1"># equivalent alternative</span>
<span class="gh">Out[5]: </span>
<span class="go">array([ 0,  0,  0,  0,  0,  1,  1,  1,  1,  1,  2,  2,  2,  2,  2,  3,  3,</span>
<span class="go">        3,  3,  3,  4,  4,  4,  4,  4,  5,  5,  5,  5,  5,  6,  6,  6,  6,</span>
<span class="go">        6,  7,  7,  7,  7,  7,  8,  8,  8,  8,  9,  9,  9,  9, 10, 10, 10,</span>
<span class="go">       10, 11, 11, 11, 11, 12, 12, 12, 12, 13, 13, 13, 13, 14, 14, 14, 14,</span>
<span class="go">       15, 15, 15, 15, 16, 16, 16, 16, 17, 17, 17, 17, 18, 18, 18, 18, 19,</span>
<span class="go">       19, 19, 19, 20, 20, 20, 20, 21, 21, 21, 21, 22, 22, 22, 22])</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Calling <code class="docutils literal notranslate"><span class="pre">values</span></code> materializes a variable’s data in memory, so is only suitable for small datasets.</p>
</div>
</section>
<section id="find-the-definition-for-a-variable-in-a-dataset">
<h2><a class="toc-backref" href="#id5" role="doc-backlink">Find the definition for a variable in a dataset?</a><a class="headerlink" href="#find-the-definition-for-a-variable-in-a-dataset" title="Permalink to this heading">#</a></h2>
<p>Use the <code class="docutils literal notranslate"><span class="pre">comment</span></code> attribute on the variable:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [6]: </span><span class="n">ds</span><span class="o">.</span><span class="n">variant_contig</span><span class="o">.</span><span class="n">comment</span>
<span class="gh">Out[6]: </span><span class="go">&#39;Index corresponding to contig name for each variant. In some less common\nscenarios, this may also be equivalent to the contig names if the data\ngenerating process used contig names that were also integers.&#39;</span>
</pre></div>
</div>
<p>All the variables defined in sgkit are documented on the <a class="reference internal" href="api.html#api-variables"><span class="std std-ref">Variables</span></a> API page.</p>
</section>
<section id="look-at-the-genotypes">
<h2><a class="toc-backref" href="#id6" role="doc-backlink">Look at the genotypes?</a><a class="headerlink" href="#look-at-the-genotypes" title="Permalink to this heading">#</a></h2>
<p>Call <a class="reference internal" href="generated/sgkit.display_genotypes.html#sgkit.display_genotypes" title="sgkit.display_genotypes"><code class="xref py py-func docutils literal notranslate"><span class="pre">display_genotypes()</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [7]: </span><span class="n">sg</span><span class="o">.</span><span class="n">display_genotypes</span><span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="n">max_variants</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="gh">Out[7]: </span>
<span class="go">samples    S0   S1   S2   S3   S4  ...  S45  S46  S47  S48  S49</span>
<span class="go">variants                           ...                         </span>
<span class="go">0         0/0  1/0  1/0  0/1  1/0  ...  1/1  0/0  1/0  0/0  1/1</span>
<span class="go">1         1/1  1/0  1/.  ./0  1/0  ...  1/1  0/1  1/0  1/1  1/0</span>
<span class="go">2         0/1  1/1  1/1  1/0  1/1  ...  0/0  0/1  0/0  0/0  1/1</span>
<span class="go">3         1/1  0/0  1/1  ./1  0/1  ...  0/1  1/0  0/1  0/.  0/.</span>
<span class="go">4         1/0  0/1  0/1  0/1  0/0  ...  1/0  1/1  0/0  1/.  1/0</span>
<span class="go">...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...</span>
<span class="go">6         1/1  1/1  ./0  1/1  0/1  ...  0/0  0/.  1/0  1/0  0/1</span>
<span class="go">7         1/.  1/0  ./0  0/1  1/0  ...  0/1  1/.  0/0  1/0  0/0</span>
<span class="go">8         0/1  0/0  0/0  0/1  0/0  ...  0/1  0/1  1/0  1/0  0/0</span>
<span class="go">9         1/1  0/0  ./1  1/0  0/0  ...  0/0  0/0  1/1  0/1  1/0</span>
<span class="go">10        1/1  0/.  0/0  0/1  1/.  ...  1/0  0/.  0/1  0/1  0/0</span>

<span class="go">[100 rows x 50 columns]</span>
</pre></div>
</div>
</section>
<section id="subset-the-variables">
<h2><a class="toc-backref" href="#id7" role="doc-backlink">Subset the variables?</a><a class="headerlink" href="#subset-the-variables" title="Permalink to this heading">#</a></h2>
<p>Use Xarray’s pandas-like method for <a class="reference external" href="http://xarray.pydata.org/en/latest/user-guide/data-structures.html#transforming-datasets">selecting variables</a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [8]: </span><span class="n">ds</span><span class="p">[[</span><span class="s2">&quot;variant_contig&quot;</span><span class="p">,</span> <span class="s2">&quot;variant_position&quot;</span><span class="p">,</span> <span class="s2">&quot;variant_allele&quot;</span><span class="p">]]</span>
<span class="gh">Out[8]: </span>
<span class="go">&lt;xarray.Dataset&gt; Size: 2kB</span>
<span class="go">Dimensions:           (variants: 100, alleles: 2)</span>
<span class="go">Dimensions without coordinates: variants, alleles</span>
<span class="go">Data variables:</span>
<span class="go">    variant_contig    (variants) int64 800B 0 0 0 0 0 1 1 ... 21 21 22 22 22 22</span>
<span class="go">    variant_position  (variants) int64 800B 0 1 2 3 4 0 1 2 ... 0 1 2 3 0 1 2 3</span>
<span class="go">    variant_allele    (variants, alleles) |S1 200B b&#39;T&#39; b&#39;C&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [&#39;0&#39;, &#39;1&#39;, &#39;2&#39;, &#39;3&#39;, &#39;4&#39;, &#39;5&#39;, &#39;6&#39;, &#39;7&#39;, &#39;8&#39;, &#39;9&#39;, &#39;10&#39;, &#39;11&#39;, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
<p>Alternatively, you can <a class="reference external" href="http://xarray.pydata.org/en/latest/generated/xarray.Dataset.drop_vars.html#xarray.Dataset.drop_vars">drop variables</a> that you want to remove:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [9]: </span><span class="n">ds</span><span class="o">.</span><span class="n">drop_vars</span><span class="p">([</span><span class="s2">&quot;variant_contig&quot;</span><span class="p">,</span> <span class="s2">&quot;variant_position&quot;</span><span class="p">,</span> <span class="s2">&quot;variant_allele&quot;</span><span class="p">])</span>
<span class="gh">Out[9]: </span>
<span class="go">&lt;xarray.Dataset&gt; Size: 21kB</span>
<span class="go">Dimensions:             (contigs: 23, samples: 50, variants: 100, ploidy: 2)</span>
<span class="go">Dimensions without coordinates: contigs, samples, variants, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    contig_id           (contigs) &lt;U2 184B &#39;0&#39; &#39;1&#39; &#39;2&#39; &#39;3&#39; ... &#39;20&#39; &#39;21&#39; &#39;22&#39;</span>
<span class="go">    sample_id           (samples) &lt;U3 600B &#39;S0&#39; &#39;S1&#39; &#39;S2&#39; ... &#39;S47&#39; &#39;S48&#39; &#39;S49&#39;</span>
<span class="go">    call_genotype       (variants, samples, ploidy) int8 10kB 0 0 1 0 ... 1 0 0</span>
<span class="go">    call_genotype_mask  (variants, samples, ploidy) bool 10kB False ... False</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [&#39;0&#39;, &#39;1&#39;, &#39;2&#39;, &#39;3&#39;, &#39;4&#39;, &#39;5&#39;, &#39;6&#39;, &#39;7&#39;, &#39;8&#39;, &#39;9&#39;, &#39;10&#39;, &#39;11&#39;, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
</section>
<section id="subset-to-a-genomic-range">
<h2><a class="toc-backref" href="#id8" role="doc-backlink">Subset to a genomic range?</a><a class="headerlink" href="#subset-to-a-genomic-range" title="Permalink to this heading">#</a></h2>
<p>Set an index on the dataset, then call <a class="reference external" href="https://docs.xarray.dev/en/stable/generated/xarray.Dataset.sel.html#xarray.Dataset.sel" title="(in xarray v2025.1.2.dev0)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">xarray.Dataset.sel()</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [10]: </span><span class="n">ds</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">variants</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;variant_contig&quot;</span><span class="p">,</span> <span class="s2">&quot;variant_position&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">sel</span><span class="p">(</span><span class="n">variants</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">slice</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">)))</span>
<span class="gh">Out[10]: </span>
<span class="go">&lt;xarray.Dataset&gt; Size: 1kB</span>
<span class="go">Dimensions:             (contigs: 23, variants: 3, alleles: 2, samples: 50,</span>
<span class="go">                         ploidy: 2)</span>
<span class="go">Coordinates:</span>
<span class="go">  * variants            (variants) object 24B MultiIndex</span>
<span class="go">  * variant_contig      (variants) int64 24B 0 0 0</span>
<span class="go">  * variant_position    (variants) int64 24B 2 3 4</span>
<span class="go">Dimensions without coordinates: contigs, alleles, samples, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    contig_id           (contigs) &lt;U2 184B &#39;0&#39; &#39;1&#39; &#39;2&#39; &#39;3&#39; ... &#39;20&#39; &#39;21&#39; &#39;22&#39;</span>
<span class="go">    variant_allele      (variants, alleles) |S1 6B b&#39;T&#39; b&#39;G&#39; b&#39;G&#39; b&#39;G&#39; b&#39;C&#39; b&#39;G&#39;</span>
<span class="go">    sample_id           (samples) &lt;U3 600B &#39;S0&#39; &#39;S1&#39; &#39;S2&#39; ... &#39;S47&#39; &#39;S48&#39; &#39;S49&#39;</span>
<span class="go">    call_genotype       (variants, samples, ploidy) int8 300B 0 1 1 1 ... -1 1 0</span>
<span class="go">    call_genotype_mask  (variants, samples, ploidy) bool 300B False ... False</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [&#39;0&#39;, &#39;1&#39;, &#39;2&#39;, &#39;3&#39;, &#39;4&#39;, &#39;5&#39;, &#39;6&#39;, &#39;7&#39;, &#39;8&#39;, &#39;9&#39;, &#39;10&#39;, &#39;11&#39;, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
<p>An API to make this easier is under discussion. Please add your requirements to <a class="github reference external" href="https://github.com/sgkit-dev/sgkit/pull/658">sgkit-dev/sgkit#658</a>.</p>
</section>
<section id="get-the-list-of-samples">
<h2><a class="toc-backref" href="#id9" role="doc-backlink">Get the list of samples?</a><a class="headerlink" href="#get-the-list-of-samples" title="Permalink to this heading">#</a></h2>
<p>Get the values for the <code class="docutils literal notranslate"><span class="pre">sample_id</span></code> variable:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [11]: </span><span class="n">ds</span><span class="o">.</span><span class="n">sample_id</span><span class="o">.</span><span class="n">values</span>
<span class="gh">Out[11]: </span>
<span class="go">array([&#39;S0&#39;, &#39;S1&#39;, &#39;S2&#39;, &#39;S3&#39;, &#39;S4&#39;, &#39;S5&#39;, &#39;S6&#39;, &#39;S7&#39;, &#39;S8&#39;, &#39;S9&#39;, &#39;S10&#39;,</span>
<span class="go">       &#39;S11&#39;, &#39;S12&#39;, &#39;S13&#39;, &#39;S14&#39;, &#39;S15&#39;, &#39;S16&#39;, &#39;S17&#39;, &#39;S18&#39;, &#39;S19&#39;,</span>
<span class="go">       &#39;S20&#39;, &#39;S21&#39;, &#39;S22&#39;, &#39;S23&#39;, &#39;S24&#39;, &#39;S25&#39;, &#39;S26&#39;, &#39;S27&#39;, &#39;S28&#39;,</span>
<span class="go">       &#39;S29&#39;, &#39;S30&#39;, &#39;S31&#39;, &#39;S32&#39;, &#39;S33&#39;, &#39;S34&#39;, &#39;S35&#39;, &#39;S36&#39;, &#39;S37&#39;,</span>
<span class="go">       &#39;S38&#39;, &#39;S39&#39;, &#39;S40&#39;, &#39;S41&#39;, &#39;S42&#39;, &#39;S43&#39;, &#39;S44&#39;, &#39;S45&#39;, &#39;S46&#39;,</span>
<span class="go">       &#39;S47&#39;, &#39;S48&#39;, &#39;S49&#39;], dtype=&#39;&lt;U3&#39;)</span>
</pre></div>
</div>
</section>
<section id="subset-the-samples">
<h2><a class="toc-backref" href="#id10" role="doc-backlink">Subset the samples?</a><a class="headerlink" href="#subset-the-samples" title="Permalink to this heading">#</a></h2>
<p>Call <a class="reference external" href="https://docs.xarray.dev/en/stable/generated/xarray.Dataset.sel.html#xarray.Dataset.sel" title="(in xarray v2025.1.2.dev0)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">xarray.Dataset.sel()</span></code></a> and <a class="reference external" href="https://docs.xarray.dev/en/stable/generated/xarray.DataArray.isin.html#xarray.DataArray.isin" title="(in xarray v2025.1.2.dev0)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">xarray.DataArray.isin()</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [12]: </span><span class="n">ds</span><span class="o">.</span><span class="n">sel</span><span class="p">(</span><span class="n">samples</span><span class="o">=</span><span class="n">ds</span><span class="o">.</span><span class="n">sample_id</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;S30&quot;</span><span class="p">,</span> <span class="s2">&quot;S32&quot;</span><span class="p">]))</span>
<span class="gh">Out[12]: </span>
<span class="go">&lt;xarray.Dataset&gt; Size: 3kB</span>
<span class="go">Dimensions:             (contigs: 23, variants: 100, alleles: 2, samples: 2,</span>
<span class="go">                         ploidy: 2)</span>
<span class="go">Dimensions without coordinates: contigs, variants, alleles, samples, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    contig_id           (contigs) &lt;U2 184B &#39;0&#39; &#39;1&#39; &#39;2&#39; &#39;3&#39; ... &#39;20&#39; &#39;21&#39; &#39;22&#39;</span>
<span class="go">    variant_contig      (variants) int64 800B 0 0 0 0 0 1 ... 21 21 22 22 22 22</span>
<span class="go">    variant_position    (variants) int64 800B 0 1 2 3 4 0 1 2 ... 1 2 3 0 1 2 3</span>
<span class="go">    variant_allele      (variants, alleles) |S1 200B b&#39;T&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    sample_id           (samples) &lt;U3 24B &#39;S30&#39; &#39;S32&#39;</span>
<span class="go">    call_genotype       (variants, samples, ploidy) int8 400B 0 -1 0 0 ... 1 0 0</span>
<span class="go">    call_genotype_mask  (variants, samples, ploidy) bool 400B False ... False</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [&#39;0&#39;, &#39;1&#39;, &#39;2&#39;, &#39;3&#39;, &#39;4&#39;, &#39;5&#39;, &#39;6&#39;, &#39;7&#39;, &#39;8&#39;, &#39;9&#39;, &#39;10&#39;, &#39;11&#39;, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
</section>
<section id="define-a-new-variable-based-on-others">
<h2><a class="toc-backref" href="#id11" role="doc-backlink">Define a new variable based on others?</a><a class="headerlink" href="#define-a-new-variable-based-on-others" title="Permalink to this heading">#</a></h2>
<p>Use Xarray’s <a class="reference external" href="http://xarray.pydata.org/en/stable/user-guide/data-structures.html#dictionary-like-methods">dictionary like methods</a>, or <a class="reference external" href="https://docs.xarray.dev/en/stable/generated/xarray.Dataset.assign.html#xarray.Dataset.assign" title="(in xarray v2025.1.2.dev0)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">xarray.Dataset.assign()</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [13]: </span><span class="n">ds</span><span class="p">[</span><span class="s2">&quot;pos0&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">variant_position</span> <span class="o">-</span> <span class="mi">1</span>

<span class="gp">In [14]: </span><span class="n">ds</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">pos0</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">variant_position</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="c1"># alternative</span>
<span class="gh">Out[14]: </span>
<span class="go">&lt;xarray.Dataset&gt; Size: 23kB</span>
<span class="go">Dimensions:             (contigs: 23, variants: 100, alleles: 2, samples: 50,</span>
<span class="go">                         ploidy: 2)</span>
<span class="go">Dimensions without coordinates: contigs, variants, alleles, samples, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    contig_id           (contigs) &lt;U2 184B &#39;0&#39; &#39;1&#39; &#39;2&#39; &#39;3&#39; ... &#39;20&#39; &#39;21&#39; &#39;22&#39;</span>
<span class="go">    variant_contig      (variants) int64 800B 0 0 0 0 0 1 ... 21 21 22 22 22 22</span>
<span class="go">    variant_position    (variants) int64 800B 0 1 2 3 4 0 1 2 ... 1 2 3 0 1 2 3</span>
<span class="go">    variant_allele      (variants, alleles) |S1 200B b&#39;T&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    sample_id           (samples) &lt;U3 600B &#39;S0&#39; &#39;S1&#39; &#39;S2&#39; ... &#39;S47&#39; &#39;S48&#39; &#39;S49&#39;</span>
<span class="go">    call_genotype       (variants, samples, ploidy) int8 10kB 0 0 1 0 ... 1 0 0</span>
<span class="go">    call_genotype_mask  (variants, samples, ploidy) bool 10kB False ... False</span>
<span class="go">    pos0                (variants) int64 800B -1 0 1 2 3 -1 0 ... 0 1 2 -1 0 1 2</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [&#39;0&#39;, &#39;1&#39;, &#39;2&#39;, &#39;3&#39;, &#39;4&#39;, &#39;5&#39;, &#39;6&#39;, &#39;7&#39;, &#39;8&#39;, &#39;9&#39;, &#39;10&#39;, &#39;11&#39;, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
</section>
<section id="get-summary-stats">
<h2><a class="toc-backref" href="#id12" role="doc-backlink">Get summary stats?</a><a class="headerlink" href="#get-summary-stats" title="Permalink to this heading">#</a></h2>
<p>Call <a class="reference internal" href="generated/sgkit.sample_stats.html#sgkit.sample_stats" title="sgkit.sample_stats"><code class="xref py py-func docutils literal notranslate"><span class="pre">sample_stats()</span></code></a> or <a class="reference internal" href="generated/sgkit.variant_stats.html#sgkit.variant_stats" title="sgkit.variant_stats"><code class="xref py py-func docutils literal notranslate"><span class="pre">variant_stats()</span></code></a> as appropriate:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [15]: </span><span class="n">sg</span><span class="o">.</span><span class="n">sample_stats</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>
<span class="gh">Out[15]: </span>
<span class="go">&lt;xarray.Dataset&gt; Size: 26kB</span>
<span class="go">Dimensions:             (samples: 50, contigs: 23, variants: 100, alleles: 2,</span>
<span class="go">                         ploidy: 2)</span>
<span class="go">Dimensions without coordinates: samples, contigs, variants, alleles, ploidy</span>
<span class="go">Data variables: (12/14)</span>
<span class="go">    sample_n_called     (samples) int64 400B dask.array&lt;chunksize=(50,), meta=np.ndarray&gt;</span>
<span class="go">    sample_call_rate    (samples) float64 400B dask.array&lt;chunksize=(50,), meta=np.ndarray&gt;</span>
<span class="go">    sample_n_het        (samples) int64 400B dask.array&lt;chunksize=(50,), meta=np.ndarray&gt;</span>
<span class="go">    sample_n_hom_ref    (samples) int64 400B dask.array&lt;chunksize=(50,), meta=np.ndarray&gt;</span>
<span class="go">    sample_n_hom_alt    (samples) int64 400B dask.array&lt;chunksize=(50,), meta=np.ndarray&gt;</span>
<span class="go">    sample_n_non_ref    (samples) int64 400B dask.array&lt;chunksize=(50,), meta=np.ndarray&gt;</span>
<span class="go">    ...                  ...</span>
<span class="go">    variant_position    (variants) int64 800B 0 1 2 3 4 0 1 2 ... 1 2 3 0 1 2 3</span>
<span class="go">    variant_allele      (variants, alleles) |S1 200B b&#39;T&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    sample_id           (samples) &lt;U3 600B &#39;S0&#39; &#39;S1&#39; &#39;S2&#39; ... &#39;S47&#39; &#39;S48&#39; &#39;S49&#39;</span>
<span class="go">    call_genotype       (variants, samples, ploidy) int8 10kB 0 0 1 0 ... 1 0 0</span>
<span class="go">    call_genotype_mask  (variants, samples, ploidy) bool 10kB False ... False</span>
<span class="go">    pos0                (variants) int64 800B -1 0 1 2 3 -1 0 ... 0 1 2 -1 0 1 2</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [&#39;0&#39;, &#39;1&#39;, &#39;2&#39;, &#39;3&#39;, &#39;4&#39;, &#39;5&#39;, &#39;6&#39;, &#39;7&#39;, &#39;8&#39;, &#39;9&#39;, &#39;10&#39;, &#39;11&#39;, ...</span>
<span class="go">    source:   sgkit-unknown</span>

<span class="gp">In [16]: </span><span class="n">sg</span><span class="o">.</span><span class="n">variant_stats</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>
<span class="gh">Out[16]: </span>
<span class="go">&lt;xarray.Dataset&gt; Size: 32kB</span>
<span class="go">Dimensions:                   (variants: 100, alleles: 2, contigs: 23,</span>
<span class="go">                               samples: 50, ploidy: 2)</span>
<span class="go">Dimensions without coordinates: variants, alleles, contigs, samples, ploidy</span>
<span class="go">Data variables: (12/17)</span>
<span class="go">    variant_n_called          (variants) int64 800B dask.array&lt;chunksize=(100,), meta=np.ndarray&gt;</span>
<span class="go">    variant_call_rate         (variants) float64 800B dask.array&lt;chunksize=(100,), meta=np.ndarray&gt;</span>
<span class="go">    variant_n_het             (variants) int64 800B dask.array&lt;chunksize=(100,), meta=np.ndarray&gt;</span>
<span class="go">    variant_n_hom_ref         (variants) int64 800B dask.array&lt;chunksize=(100,), meta=np.ndarray&gt;</span>
<span class="go">    variant_n_hom_alt         (variants) int64 800B dask.array&lt;chunksize=(100,), meta=np.ndarray&gt;</span>
<span class="go">    variant_n_non_ref         (variants) int64 800B dask.array&lt;chunksize=(100,), meta=np.ndarray&gt;</span>
<span class="go">    ...                        ...</span>
<span class="go">    variant_position          (variants) int64 800B 0 1 2 3 4 0 ... 2 3 0 1 2 3</span>
<span class="go">    variant_allele            (variants, alleles) |S1 200B b&#39;T&#39; b&#39;C&#39; ... b&#39;A&#39;</span>
<span class="go">    sample_id                 (samples) &lt;U3 600B &#39;S0&#39; &#39;S1&#39; &#39;S2&#39; ... &#39;S48&#39; &#39;S49&#39;</span>
<span class="go">    call_genotype             (variants, samples, ploidy) int8 10kB 0 0 ... 0 0</span>
<span class="go">    call_genotype_mask        (variants, samples, ploidy) bool 10kB False ......</span>
<span class="go">    pos0                      (variants) int64 800B -1 0 1 2 3 -1 ... 2 -1 0 1 2</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [&#39;0&#39;, &#39;1&#39;, &#39;2&#39;, &#39;3&#39;, &#39;4&#39;, &#39;5&#39;, &#39;6&#39;, &#39;7&#39;, &#39;8&#39;, &#39;9&#39;, &#39;10&#39;, &#39;11&#39;, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
</section>
<section id="filter-variants">
<h2><a class="toc-backref" href="#id13" role="doc-backlink">Filter variants?</a><a class="headerlink" href="#filter-variants" title="Permalink to this heading">#</a></h2>
<p>Call <a class="reference external" href="https://docs.xarray.dev/en/stable/generated/xarray.Dataset.sel.html#xarray.Dataset.sel" title="(in xarray v2025.1.2.dev0)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">xarray.Dataset.sel()</span></code></a> on the <code class="docutils literal notranslate"><span class="pre">variants</span></code> dimension:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [17]: </span><span class="n">ds2</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">hardy_weinberg_test</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>

<span class="gp">In [18]: </span><span class="n">ds2</span><span class="o">.</span><span class="n">sel</span><span class="p">(</span><span class="n">variants</span><span class="o">=</span><span class="p">(</span><span class="n">ds2</span><span class="o">.</span><span class="n">variant_hwe_p_value</span> <span class="o">&gt;</span> <span class="mf">1e-2</span><span class="p">)</span><span class="o">.</span><span class="n">compute</span><span class="p">())</span>
<span class="gh">Out[18]: </span>
<span class="go">&lt;xarray.Dataset&gt; Size: 26kB</span>
<span class="go">Dimensions:                 (variants: 99, genotypes: 3, contigs: 23,</span>
<span class="go">                             alleles: 2, samples: 50, ploidy: 2)</span>
<span class="go">Coordinates:</span>
<span class="go">  * genotypes               (genotypes) &lt;U3 36B &#39;0/0&#39; &#39;0/1&#39; &#39;1/1&#39;</span>
<span class="go">Dimensions without coordinates: variants, contigs, alleles, samples, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    variant_hwe_p_value     (variants) float64 792B dask.array&lt;chunksize=(99,), meta=np.ndarray&gt;</span>
<span class="go">    variant_genotype_count  (variants, genotypes) uint64 2kB dask.array&lt;chunksize=(99, 3), meta=np.ndarray&gt;</span>
<span class="go">    genotype_id             (genotypes) &lt;U3 36B dask.array&lt;chunksize=(3,), meta=np.ndarray&gt;</span>
<span class="go">    contig_id               (contigs) &lt;U2 184B &#39;0&#39; &#39;1&#39; &#39;2&#39; ... &#39;20&#39; &#39;21&#39; &#39;22&#39;</span>
<span class="go">    variant_contig          (variants) int64 792B 0 0 0 0 1 1 ... 21 22 22 22 22</span>
<span class="go">    variant_position        (variants) int64 792B 0 2 3 4 0 1 2 ... 2 3 0 1 2 3</span>
<span class="go">    variant_allele          (variants, alleles) |S1 198B b&#39;T&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    sample_id               (samples) &lt;U3 600B &#39;S0&#39; &#39;S1&#39; &#39;S2&#39; ... &#39;S48&#39; &#39;S49&#39;</span>
<span class="go">    call_genotype           (variants, samples, ploidy) int8 10kB 0 0 1 ... 0 0</span>
<span class="go">    call_genotype_mask      (variants, samples, ploidy) bool 10kB False ... F...</span>
<span class="go">    pos0                    (variants) int64 792B -1 1 2 3 -1 0 ... 1 2 -1 0 1 2</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [&#39;0&#39;, &#39;1&#39;, &#39;2&#39;, &#39;3&#39;, &#39;4&#39;, &#39;5&#39;, &#39;6&#39;, &#39;7&#39;, &#39;8&#39;, &#39;9&#39;, &#39;10&#39;, &#39;11&#39;, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The call to <code class="docutils literal notranslate"><span class="pre">compute</span></code> is needed to avoid an Xarray error.</p>
</div>
</section>
<section id="find-which-new-variables-were-added-by-a-method">
<h2><a class="toc-backref" href="#id14" role="doc-backlink">Find which new variables were added by a method?</a><a class="headerlink" href="#find-which-new-variables-were-added-by-a-method" title="Permalink to this heading">#</a></h2>
<p>Use <a class="reference external" href="https://docs.xarray.dev/en/stable/generated/xarray.Dataset.data_vars.html#xarray.Dataset.data_vars" title="(in xarray v2025.1.2.dev0)"><code class="xref py py-attr docutils literal notranslate"><span class="pre">xarray.Dataset.data_vars</span></code></a> to compare the new dataset variables to the old:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [19]: </span><span class="n">ds2</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">sample_stats</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>

<span class="gp">In [20]: </span><span class="nb">set</span><span class="p">(</span><span class="n">ds2</span><span class="o">.</span><span class="n">data_vars</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">data_vars</span><span class="p">)</span>
<span class="gh">Out[20]: </span>
<span class="go">{&#39;sample_call_rate&#39;,</span>
<span class="go"> &#39;sample_n_called&#39;,</span>
<span class="go"> &#39;sample_n_het&#39;,</span>
<span class="go"> &#39;sample_n_hom_alt&#39;,</span>
<span class="go"> &#39;sample_n_hom_ref&#39;,</span>
<span class="go"> &#39;sample_n_non_ref&#39;}</span>
</pre></div>
</div>
</section>
<section id="save-results-to-a-zarr-file">
<h2><a class="toc-backref" href="#id15" role="doc-backlink">Save results to a Zarr file?</a><a class="headerlink" href="#save-results-to-a-zarr-file" title="Permalink to this heading">#</a></h2>
<p>Call <a class="reference internal" href="generated/sgkit.save_dataset.html#sgkit.save_dataset" title="sgkit.save_dataset"><code class="xref py py-func docutils literal notranslate"><span class="pre">save_dataset()</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [21]: </span><span class="n">sg</span><span class="o">.</span><span class="n">save_dataset</span><span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="s2">&quot;ds.zarr&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Zarr datasets must have equal-sized chunks (except for the final chunk, which may be smaller),
so you may have to <a class="reference external" href="http://xarray.pydata.org/en/stable/generated/xarray.DataArray.chunk.html">rechunk the dataset</a> first.</p>
</div>
</section>
<section id="load-a-dataset-from-zarr">
<h2><a class="toc-backref" href="#id16" role="doc-backlink">Load a dataset from Zarr?</a><a class="headerlink" href="#load-a-dataset-from-zarr" title="Permalink to this heading">#</a></h2>
<p>Call <a class="reference internal" href="generated/sgkit.load_dataset.html#sgkit.load_dataset" title="sgkit.load_dataset"><code class="xref py py-func docutils literal notranslate"><span class="pre">load_dataset()</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [22]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="s2">&quot;ds.zarr&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>

<div class="section ablog__blog_comments">
   
</div>

                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="generated/sgkit.variables.window_stop_spec.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">sgkit.variables.window_stop_spec</p>
      </div>
    </a>
    <a class="right-next"
       href="contributing.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Contributing to sgkit</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#create-a-test-dataset">Create a test dataset?</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#look-at-the-dataset-summary">Look at the dataset summary?</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#get-the-values-for-a-variable-in-a-dataset">Get the values for a variable in a dataset?</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#find-the-definition-for-a-variable-in-a-dataset">Find the definition for a variable in a dataset?</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#look-at-the-genotypes">Look at the genotypes?</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#subset-the-variables">Subset the variables?</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#subset-to-a-genomic-range">Subset to a genomic range?</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#get-the-list-of-samples">Get the list of samples?</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#subset-the-samples">Subset the samples?</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#define-a-new-variable-based-on-others">Define a new variable based on others?</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#get-summary-stats">Get summary stats?</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#filter-variants">Filter variants?</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#find-which-new-variables-were-added-by-a-method">Find which new variables were added by a method?</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#save-results-to-a-zarr-file">Save results to a Zarr file?</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#load-a-dataset-from-zarr">Load a dataset from Zarr?</a></li>
</ul>
  </nav></div>

  <div class="sidebar-secondary-item">
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/how_do_i.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2020, sgkit developers.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 6.2.1.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>