
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>How do I …</title>
    
  <link href="_static/css/theme.css" rel="stylesheet">
  <link href="_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="_static/typehints.css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/docsearch.sbt.css" />
    <link rel="stylesheet" type="text/css" href="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.css" />
    
  <link rel="preload" as="script" href="_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/sphinx-book-theme.d59cb220de22ca1c485ebbdc042f0030.js"></script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.js"></script>
    <script defer="defer" src="_static/docsearch.sbt.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Contributing to sgkit" href="contributing.html" />
    <link rel="prev" title="sgkit.variables.window_stop_spec" href="generated/sgkit.variables.window_stop_spec.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/sgkit_trnsprnt.png" class="logo" alt="logo">
      
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Contents
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="getting_started.html">
   Getting Started
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="user_guide.html">
   User Guide
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="vcf.html">
   Reading VCF
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="examples/index.html">
   Examples
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="examples/gwas_tutorial.html">
     GWAS Tutorial
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="api.html">
   API reference
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.io.bgen.bgen_to_zarr.html">
     sgkit.io.bgen.bgen_to_zarr
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.io.bgen.read_bgen.html">
     sgkit.io.bgen.read_bgen
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.io.bgen.rechunk_bgen.html">
     sgkit.io.bgen.rechunk_bgen
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.io.plink.read_plink.html">
     sgkit.io.plink.read_plink
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.io.vcf.vcf_to_zarr.html">
     sgkit.io.vcf.vcf_to_zarr
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.io.vcf.partition_into_regions.html">
     sgkit.io.vcf.partition_into_regions
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.io.vcf.vcf_to_zarrs.html">
     sgkit.io.vcf.vcf_to_zarrs
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.io.vcf.concat_zarrs.html">
     sgkit.io.vcf.concat_zarrs
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.read_scikit_allel_vcfzarr.html">
     sgkit.read_scikit_allel_vcfzarr
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.load_dataset.html">
     sgkit.load_dataset
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.save_dataset.html">
     sgkit.save_dataset
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.call_allele_frequencies.html">
     sgkit.call_allele_frequencies
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.count_call_alleles.html">
     sgkit.count_call_alleles
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.count_cohort_alleles.html">
     sgkit.count_cohort_alleles
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.count_variant_alleles.html">
     sgkit.count_variant_alleles
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.divergence.html">
     sgkit.divergence
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.diversity.html">
     sgkit.diversity
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.Fst.html">
     sgkit.Fst
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.Garud_H.html">
     sgkit.Garud_H
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.gwas_linear_regression.html">
     sgkit.gwas_linear_regression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.hardy_weinberg_test.html">
     sgkit.hardy_weinberg_test
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.identity_by_state.html">
     sgkit.identity_by_state
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.individual_heterozygosity.html">
     sgkit.individual_heterozygosity
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.ld_matrix.html">
     sgkit.ld_matrix
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.ld_prune.html">
     sgkit.ld_prune
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.maximal_independent_set.html">
     sgkit.maximal_independent_set
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.observed_heterozygosity.html">
     sgkit.observed_heterozygosity
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.pbs.html">
     sgkit.pbs
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.pc_relate.html">
     sgkit.pc_relate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.regenie.html">
     sgkit.regenie
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.sample_stats.html">
     sgkit.sample_stats
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.Tajimas_D.html">
     sgkit.Tajimas_D
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variant_stats.html">
     sgkit.variant_stats
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.Weir_Goudet_beta.html">
     sgkit.Weir_Goudet_beta
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.convert_probability_to_call.html">
     sgkit.convert_probability_to_call
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.display_genotypes.html">
     sgkit.display_genotypes
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.filter_partial_calls.html">
     sgkit.filter_partial_calls
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.infer_call_ploidy.html">
     sgkit.infer_call_ploidy
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.infer_sample_ploidy.html">
     sgkit.infer_sample_ploidy
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.infer_variant_ploidy.html">
     sgkit.infer_variant_ploidy
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.simulate_genotype_call_dataset.html">
     sgkit.simulate_genotype_call_dataset
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.window_by_position.html">
     sgkit.window_by_position
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.window_by_variant.html">
     sgkit.window_by_variant
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.regenie_base_prediction_spec.html">
     sgkit.variables.regenie_base_prediction_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_allele_count_spec.html">
     sgkit.variables.call_allele_count_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_allele_frequency_spec.html">
     sgkit.variables.call_allele_frequency_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_dosage_spec.html">
     sgkit.variables.call_dosage_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_dosage_mask_spec.html">
     sgkit.variables.call_dosage_mask_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_genotype_complete_spec.html">
     sgkit.variables.call_genotype_complete_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_genotype_complete_mask_spec.html">
     sgkit.variables.call_genotype_complete_mask_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_genotype_spec.html">
     sgkit.variables.call_genotype_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_genotype_mask_spec.html">
     sgkit.variables.call_genotype_mask_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_genotype_fill_spec.html">
     sgkit.variables.call_genotype_fill_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_genotype_phased_spec.html">
     sgkit.variables.call_genotype_phased_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_genotype_probability_spec.html">
     sgkit.variables.call_genotype_probability_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_genotype_probability_mask_spec.html">
     sgkit.variables.call_genotype_probability_mask_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_heterozygosity_spec.html">
     sgkit.variables.call_heterozygosity_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.call_ploidy_spec.html">
     sgkit.variables.call_ploidy_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.cohort_allele_count_spec.html">
     sgkit.variables.cohort_allele_count_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.covariates_spec.html">
     sgkit.variables.covariates_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.dosage_spec.html">
     sgkit.variables.dosage_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.genotype_count_spec.html">
     sgkit.variables.genotype_count_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.ld_prune_index_to_drop_spec.html">
     sgkit.variables.ld_prune_index_to_drop_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.regenie_loco_prediction_spec.html">
     sgkit.variables.regenie_loco_prediction_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.regenie_meta_prediction_spec.html">
     sgkit.variables.regenie_meta_prediction_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.pc_relate_phi_spec.html">
     sgkit.variables.pc_relate_phi_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_call_rate_spec.html">
     sgkit.variables.sample_call_rate_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_cohort_spec.html">
     sgkit.variables.sample_cohort_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_id_spec.html">
     sgkit.variables.sample_id_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_n_called_spec.html">
     sgkit.variables.sample_n_called_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_n_het_spec.html">
     sgkit.variables.sample_n_het_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_n_hom_alt_spec.html">
     sgkit.variables.sample_n_hom_alt_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_n_hom_ref_spec.html">
     sgkit.variables.sample_n_hom_ref_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_n_non_ref_spec.html">
     sgkit.variables.sample_n_non_ref_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_pca_component_spec.html">
     sgkit.variables.sample_pca_component_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_pca_explained_variance_spec.html">
     sgkit.variables.sample_pca_explained_variance_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_pca_explained_variance_ratio_spec.html">
     sgkit.variables.sample_pca_explained_variance_ratio_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_pca_loading_spec.html">
     sgkit.variables.sample_pca_loading_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_pca_projection_spec.html">
     sgkit.variables.sample_pca_projection_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.sample_ploidy_spec.html">
     sgkit.variables.sample_ploidy_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_divergence_spec.html">
     sgkit.variables.stat_divergence_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_diversity_spec.html">
     sgkit.variables.stat_diversity_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_Fst_spec.html">
     sgkit.variables.stat_Fst_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_Garud_h1_spec.html">
     sgkit.variables.stat_Garud_h1_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_Garud_h12_spec.html">
     sgkit.variables.stat_Garud_h12_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_Garud_h123_spec.html">
     sgkit.variables.stat_Garud_h123_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_Garud_h2_h1_spec.html">
     sgkit.variables.stat_Garud_h2_h1_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_identity_by_state_spec.html">
     sgkit.variables.stat_identity_by_state_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_observed_heterozygosity_spec.html">
     sgkit.variables.stat_observed_heterozygosity_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_pbs_spec.html">
     sgkit.variables.stat_pbs_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_Tajimas_D_spec.html">
     sgkit.variables.stat_Tajimas_D_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.stat_Weir_Goudet_beta_spec.html">
     sgkit.variables.stat_Weir_Goudet_beta_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.traits_spec.html">
     sgkit.variables.traits_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_allele_spec.html">
     sgkit.variables.variant_allele_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_allele_count_spec.html">
     sgkit.variables.variant_allele_count_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_allele_frequency_spec.html">
     sgkit.variables.variant_allele_frequency_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_allele_total_spec.html">
     sgkit.variables.variant_allele_total_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_linreg_beta_spec.html">
     sgkit.variables.variant_linreg_beta_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_call_rate_spec.html">
     sgkit.variables.variant_call_rate_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_contig_spec.html">
     sgkit.variables.variant_contig_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_hwe_p_value_spec.html">
     sgkit.variables.variant_hwe_p_value_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_id_spec.html">
     sgkit.variables.variant_id_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_n_called_spec.html">
     sgkit.variables.variant_n_called_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_n_het_spec.html">
     sgkit.variables.variant_n_het_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_n_hom_alt_spec.html">
     sgkit.variables.variant_n_hom_alt_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_n_hom_ref_spec.html">
     sgkit.variables.variant_n_hom_ref_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_n_non_ref_spec.html">
     sgkit.variables.variant_n_non_ref_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_linreg_p_value_spec.html">
     sgkit.variables.variant_linreg_p_value_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_ploidy_spec.html">
     sgkit.variables.variant_ploidy_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_position_spec.html">
     sgkit.variables.variant_position_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_score_spec.html">
     sgkit.variables.variant_score_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.variant_linreg_t_value_spec.html">
     sgkit.variables.variant_linreg_t_value_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.window_contig_spec.html">
     sgkit.variables.window_contig_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.window_start_spec.html">
     sgkit.variables.window_start_spec
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="generated/sgkit.variables.window_stop_spec.html">
     sgkit.variables.window_stop_spec
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   How do I …
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="contributing.html">
   Contributing to sgkit
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="about.html">
   About
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Theme by the <a href="https://ebp.jupyterbook.org">Executable Book Project</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="_sources/how_do_i.rst.txt"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.rst</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
                onclick="printPdf(this)" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show noprint">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#create-a-test-dataset">
   Create a test dataset?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#look-at-the-dataset-summary">
   Look at the dataset summary?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#get-the-values-for-a-variable-in-a-dataset">
   Get the values for a variable in a dataset?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#find-the-definition-for-a-variable-in-a-dataset">
   Find the definition for a variable in a dataset?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#look-at-the-genotypes">
   Look at the genotypes?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#subset-the-variables">
   Subset the variables?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#subset-to-a-genomic-range">
   Subset to a genomic range?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#get-the-list-of-samples">
   Get the list of samples?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#subset-the-samples">
   Subset the samples?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#define-a-new-variable-based-on-others">
   Define a new variable based on others?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#get-summary-stats">
   Get summary stats?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#filter-variants">
   Filter variants?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#find-which-new-variables-were-added-by-a-method">
   Find which new variables were added by a method?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#save-results-to-a-zarr-file">
   Save results to a Zarr file?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#load-a-dataset-from-zarr">
   Load a dataset from Zarr?
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <!-- Table of contents that is only displayed when printing the page -->
    <div id="jb-print-docs-body" class="onlyprint">
        <h1>How do I …</h1>
        <!-- Table of contents -->
        <div id="print-main-content" class="row">
            <div class="col-12 col-md-12 pl-md-5 pr-md-5">
            <div id="jb-print-toc">
                
                <div>
                    <h2> Contents </h2>
                </div>
                <nav aria-label="Page">
                    <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#create-a-test-dataset">
   Create a test dataset?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#look-at-the-dataset-summary">
   Look at the dataset summary?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#get-the-values-for-a-variable-in-a-dataset">
   Get the values for a variable in a dataset?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#find-the-definition-for-a-variable-in-a-dataset">
   Find the definition for a variable in a dataset?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#look-at-the-genotypes">
   Look at the genotypes?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#subset-the-variables">
   Subset the variables?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#subset-to-a-genomic-range">
   Subset to a genomic range?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#get-the-list-of-samples">
   Get the list of samples?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#subset-the-samples">
   Subset the samples?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#define-a-new-variable-based-on-others">
   Define a new variable based on others?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#get-summary-stats">
   Get summary stats?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#filter-variants">
   Filter variants?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#find-which-new-variables-were-added-by-a-method">
   Find which new variables were added by a method?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#save-results-to-a-zarr-file">
   Save results to a Zarr file?
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#load-a-dataset-from-zarr">
   Load a dataset from Zarr?
  </a>
 </li>
</ul>

                </nav>
            </div>
            
              <div>
                
  <div class="section" id="how-do-i">
<span id="id1"></span><h1>How do I …<a class="headerlink" href="#how-do-i" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="contents">
<ul class="simple">
<li><p><a class="reference internal" href="#create-a-test-dataset" id="id2">Create a test dataset?</a></p></li>
<li><p><a class="reference internal" href="#look-at-the-dataset-summary" id="id3">Look at the dataset summary?</a></p></li>
<li><p><a class="reference internal" href="#get-the-values-for-a-variable-in-a-dataset" id="id4">Get the values for a variable in a dataset?</a></p></li>
<li><p><a class="reference internal" href="#find-the-definition-for-a-variable-in-a-dataset" id="id5">Find the definition for a variable in a dataset?</a></p></li>
<li><p><a class="reference internal" href="#look-at-the-genotypes" id="id6">Look at the genotypes?</a></p></li>
<li><p><a class="reference internal" href="#subset-the-variables" id="id7">Subset the variables?</a></p></li>
<li><p><a class="reference internal" href="#subset-to-a-genomic-range" id="id8">Subset to a genomic range?</a></p></li>
<li><p><a class="reference internal" href="#get-the-list-of-samples" id="id9">Get the list of samples?</a></p></li>
<li><p><a class="reference internal" href="#subset-the-samples" id="id10">Subset the samples?</a></p></li>
<li><p><a class="reference internal" href="#define-a-new-variable-based-on-others" id="id11">Define a new variable based on others?</a></p></li>
<li><p><a class="reference internal" href="#get-summary-stats" id="id12">Get summary stats?</a></p></li>
<li><p><a class="reference internal" href="#filter-variants" id="id13">Filter variants?</a></p></li>
<li><p><a class="reference internal" href="#find-which-new-variables-were-added-by-a-method" id="id14">Find which new variables were added by a method?</a></p></li>
<li><p><a class="reference internal" href="#save-results-to-a-zarr-file" id="id15">Save results to a Zarr file?</a></p></li>
<li><p><a class="reference internal" href="#load-a-dataset-from-zarr" id="id16">Load a dataset from Zarr?</a></p></li>
</ul>
</div>
<div class="section" id="create-a-test-dataset">
<h2><a class="toc-backref" href="#id2">Create a test dataset?</a><a class="headerlink" href="#create-a-test-dataset" title="Permalink to this headline">¶</a></h2>
<p>Call <a class="reference internal" href="generated/sgkit.simulate_genotype_call_dataset.html#sgkit.simulate_genotype_call_dataset" title="sgkit.simulate_genotype_call_dataset"><code class="xref py py-func docutils literal notranslate"><span class="pre">simulate_genotype_call_dataset()</span></code></a> to create a test <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v0.20.2.dev0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">xarray.Dataset</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [1]: </span><span class="kn">import</span> <span class="nn">sgkit</span> <span class="k">as</span> <span class="nn">sg</span>

<span class="gp">In [2]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">simulate_genotype_call_dataset</span><span class="p">(</span><span class="n">n_variant</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">n_sample</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">n_contig</span><span class="o">=</span><span class="mi">23</span><span class="p">,</span> <span class="n">missing_pct</span><span class="o">=</span><span class="mf">.1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="look-at-the-dataset-summary">
<h2><a class="toc-backref" href="#id3">Look at the dataset summary?</a><a class="headerlink" href="#look-at-the-dataset-summary" title="Permalink to this headline">¶</a></h2>
<p>Print using the <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v0.20.2.dev0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">xarray.Dataset</span></code></a> <code class="docutils literal notranslate"><span class="pre">repr</span></code>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [3]: </span><span class="n">ds</span>
<span class="gh">Out[3]: </span><span class="go"></span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:             (variants: 100, alleles: 2, samples: 50, ploidy: 2)</span>
<span class="go">Dimensions without coordinates: variants, alleles, samples, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    variant_contig      (variants) int64 0 0 0 0 0 1 1 ... 21 21 21 22 22 22 22</span>
<span class="go">    variant_position    (variants) int64 0 1 2 3 4 0 1 2 3 ... 3 0 1 2 3 0 1 2 3</span>
<span class="go">    variant_allele      (variants, alleles) |S1 b&#39;T&#39; b&#39;C&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    sample_id           (samples) &lt;U3 &#39;S0&#39; &#39;S1&#39; &#39;S2&#39; &#39;S3&#39; ... &#39;S47&#39; &#39;S48&#39; &#39;S49&#39;</span>
<span class="go">    call_genotype       (variants, samples, ploidy) int8 0 0 1 0 1 ... 1 0 1 0 0</span>
<span class="go">    call_genotype_mask  (variants, samples, ploidy) bool False False ... False</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
</div>
<div class="section" id="get-the-values-for-a-variable-in-a-dataset">
<h2><a class="toc-backref" href="#id4">Get the values for a variable in a dataset?</a><a class="headerlink" href="#get-the-values-for-a-variable-in-a-dataset" title="Permalink to this headline">¶</a></h2>
<p>Call <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Variable.values.html#xarray.Variable.values" title="(in xarray v0.20.2.dev0)"><code class="xref py py-attr docutils literal notranslate"><span class="pre">xarray.Variable.values</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [4]: </span><span class="n">ds</span><span class="o">.</span><span class="n">variant_contig</span><span class="o">.</span><span class="n">values</span>
<span class="gh">Out[4]: </span><span class="go"></span>
<span class="go">array([ 0,  0,  0,  0,  0,  1,  1,  1,  1,  1,  2,  2,  2,  2,  2,  3,  3,</span>
<span class="go">        3,  3,  3,  4,  4,  4,  4,  4,  5,  5,  5,  5,  5,  6,  6,  6,  6,</span>
<span class="go">        6,  7,  7,  7,  7,  7,  8,  8,  8,  8,  9,  9,  9,  9, 10, 10, 10,</span>
<span class="go">       10, 11, 11, 11, 11, 12, 12, 12, 12, 13, 13, 13, 13, 14, 14, 14, 14,</span>
<span class="go">       15, 15, 15, 15, 16, 16, 16, 16, 17, 17, 17, 17, 18, 18, 18, 18, 19,</span>
<span class="go">       19, 19, 19, 20, 20, 20, 20, 21, 21, 21, 21, 22, 22, 22, 22])</span>

<span class="gp">In [5]: </span><span class="n">ds</span><span class="p">[</span><span class="s2">&quot;variant_contig&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span> <span class="c1"># equivalent alternative</span>
<span class="gh">Out[5]: </span><span class="go"></span>
<span class="go">array([ 0,  0,  0,  0,  0,  1,  1,  1,  1,  1,  2,  2,  2,  2,  2,  3,  3,</span>
<span class="go">        3,  3,  3,  4,  4,  4,  4,  4,  5,  5,  5,  5,  5,  6,  6,  6,  6,</span>
<span class="go">        6,  7,  7,  7,  7,  7,  8,  8,  8,  8,  9,  9,  9,  9, 10, 10, 10,</span>
<span class="go">       10, 11, 11, 11, 11, 12, 12, 12, 12, 13, 13, 13, 13, 14, 14, 14, 14,</span>
<span class="go">       15, 15, 15, 15, 16, 16, 16, 16, 17, 17, 17, 17, 18, 18, 18, 18, 19,</span>
<span class="go">       19, 19, 19, 20, 20, 20, 20, 21, 21, 21, 21, 22, 22, 22, 22])</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Calling <code class="docutils literal notranslate"><span class="pre">values</span></code> materializes a variable’s data in memory, so is only suitable for small datasets.</p>
</div>
</div>
<div class="section" id="find-the-definition-for-a-variable-in-a-dataset">
<h2><a class="toc-backref" href="#id5">Find the definition for a variable in a dataset?</a><a class="headerlink" href="#find-the-definition-for-a-variable-in-a-dataset" title="Permalink to this headline">¶</a></h2>
<p>Use the <code class="docutils literal notranslate"><span class="pre">comment</span></code> attribute on the variable:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [6]: </span><span class="n">ds</span><span class="o">.</span><span class="n">variant_contig</span><span class="o">.</span><span class="n">comment</span>
<span class="gh">Out[6]: </span><span class="go">&#39;Index corresponding to contig name for each variant. In some less common\nscenarios, this may also be equivalent to the contig names if the data\ngenerating process used contig names that were also integers.&#39;</span>
</pre></div>
</div>
<p>All the variables defined in sgkit are documented on the <a class="reference internal" href="api.html#api-variables"><span class="std std-ref">Variables</span></a> API page.</p>
</div>
<div class="section" id="look-at-the-genotypes">
<h2><a class="toc-backref" href="#id6">Look at the genotypes?</a><a class="headerlink" href="#look-at-the-genotypes" title="Permalink to this headline">¶</a></h2>
<p>Call <a class="reference internal" href="generated/sgkit.display_genotypes.html#sgkit.display_genotypes" title="sgkit.display_genotypes"><code class="xref py py-func docutils literal notranslate"><span class="pre">display_genotypes()</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [7]: </span><span class="n">sg</span><span class="o">.</span><span class="n">display_genotypes</span><span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="n">max_variants</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="gh">Out[7]: </span><span class="go"></span>
<span class="go">samples    S0   S1   S2   S3   S4  ...  S45  S46  S47  S48  S49</span>
<span class="go">variants                           ...                         </span>
<span class="go">0         0/0  1/0  1/0  0/1  1/0  ...  1/1  0/0  1/0  0/0  1/1</span>
<span class="go">1         1/1  1/0  1/.  ./0  1/0  ...  1/1  0/1  1/0  1/1  1/0</span>
<span class="go">2         0/1  1/1  1/1  1/0  1/1  ...  0/0  0/1  0/0  0/0  1/1</span>
<span class="go">3         1/1  0/0  1/1  ./1  0/1  ...  0/1  1/0  0/1  0/.  0/.</span>
<span class="go">4         1/0  0/1  0/1  0/1  0/0  ...  1/0  1/1  0/0  1/.  1/0</span>
<span class="go">...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...</span>
<span class="go">7         1/1  1/1  ./0  1/1  0/1  ...  0/0  0/.  1/0  1/0  0/1</span>
<span class="go">8         1/.  1/0  ./0  0/1  1/0  ...  0/1  1/.  0/0  1/0  0/0</span>
<span class="go">9         0/1  0/0  0/0  0/1  0/0  ...  0/1  0/1  1/0  1/0  0/0</span>
<span class="go">10        1/1  0/0  ./1  1/0  0/0  ...  0/0  0/0  1/1  0/1  1/0</span>
<span class="go">11        1/1  0/.  0/0  0/1  1/.  ...  1/0  0/.  0/1  0/1  0/0</span>

<span class="go">[100 rows x 50 columns]</span>
</pre></div>
</div>
</div>
<div class="section" id="subset-the-variables">
<h2><a class="toc-backref" href="#id7">Subset the variables?</a><a class="headerlink" href="#subset-the-variables" title="Permalink to this headline">¶</a></h2>
<p>Use Xarray’s pandas-like method for <a class="reference external" href="http://xarray.pydata.org/en/latest/user-guide/data-structures.html#transforming-datasets">selecting variables</a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [8]: </span><span class="n">ds</span><span class="p">[[</span><span class="s2">&quot;variant_contig&quot;</span><span class="p">,</span> <span class="s2">&quot;variant_position&quot;</span><span class="p">,</span> <span class="s2">&quot;variant_allele&quot;</span><span class="p">]]</span>
<span class="gh">Out[8]: </span><span class="go"></span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:           (variants: 100, alleles: 2)</span>
<span class="go">Dimensions without coordinates: variants, alleles</span>
<span class="go">Data variables:</span>
<span class="go">    variant_contig    (variants) int64 0 0 0 0 0 1 1 1 ... 21 21 21 22 22 22 22</span>
<span class="go">    variant_position  (variants) int64 0 1 2 3 4 0 1 2 3 4 ... 3 0 1 2 3 0 1 2 3</span>
<span class="go">    variant_allele    (variants, alleles) |S1 b&#39;T&#39; b&#39;C&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
<p>Alternatively, you can <a class="reference external" href="http://xarray.pydata.org/en/latest/generated/xarray.Dataset.drop_vars.html#xarray.Dataset.drop_vars">drop variables</a> that you want to remove:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [9]: </span><span class="n">ds</span><span class="o">.</span><span class="n">drop_vars</span><span class="p">([</span><span class="s2">&quot;variant_contig&quot;</span><span class="p">,</span> <span class="s2">&quot;variant_position&quot;</span><span class="p">,</span> <span class="s2">&quot;variant_allele&quot;</span><span class="p">])</span>
<span class="gh">Out[9]: </span><span class="go"></span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:             (samples: 50, variants: 100, ploidy: 2)</span>
<span class="go">Dimensions without coordinates: samples, variants, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    sample_id           (samples) &lt;U3 &#39;S0&#39; &#39;S1&#39; &#39;S2&#39; &#39;S3&#39; ... &#39;S47&#39; &#39;S48&#39; &#39;S49&#39;</span>
<span class="go">    call_genotype       (variants, samples, ploidy) int8 0 0 1 0 1 ... 1 0 1 0 0</span>
<span class="go">    call_genotype_mask  (variants, samples, ploidy) bool False False ... False</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
</div>
<div class="section" id="subset-to-a-genomic-range">
<h2><a class="toc-backref" href="#id8">Subset to a genomic range?</a><a class="headerlink" href="#subset-to-a-genomic-range" title="Permalink to this headline">¶</a></h2>
<p>Set an index on the dataset, then call <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.sel.html#xarray.Dataset.sel" title="(in xarray v0.20.2.dev0)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">xarray.Dataset.sel()</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [10]: </span><span class="n">ds</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">variants</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;variant_contig&quot;</span><span class="p">,</span> <span class="s2">&quot;variant_position&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">sel</span><span class="p">(</span><span class="n">variants</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">slice</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">)))</span>
<span class="gh">Out[10]: </span><span class="go"></span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:             (variants: 3, alleles: 2, samples: 50, ploidy: 2)</span>
<span class="go">Coordinates:</span>
<span class="go">  * variants            (variants) MultiIndex</span>
<span class="go">  - variant_contig      (variants) int64 0 0 0</span>
<span class="go">  - variant_position    (variants) int64 2 3 4</span>
<span class="go">Dimensions without coordinates: alleles, samples, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    variant_allele      (variants, alleles) |S1 b&#39;T&#39; b&#39;G&#39; b&#39;G&#39; b&#39;G&#39; b&#39;C&#39; b&#39;G&#39;</span>
<span class="go">    sample_id           (samples) &lt;U3 &#39;S0&#39; &#39;S1&#39; &#39;S2&#39; &#39;S3&#39; ... &#39;S47&#39; &#39;S48&#39; &#39;S49&#39;</span>
<span class="go">    call_genotype       (variants, samples, ploidy) int8 0 1 1 1 1 ... 1 -1 1 0</span>
<span class="go">    call_genotype_mask  (variants, samples, ploidy) bool False False ... False</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
<p>An API to make this easier is under discussion. Please add your requirements to <a class="reference external" href="https://github.com/pystatgen/sgkit/pull/658">https://github.com/pystatgen/sgkit/pull/658</a>.</p>
</div>
<div class="section" id="get-the-list-of-samples">
<h2><a class="toc-backref" href="#id9">Get the list of samples?</a><a class="headerlink" href="#get-the-list-of-samples" title="Permalink to this headline">¶</a></h2>
<p>Get the values for the <code class="docutils literal notranslate"><span class="pre">sample_id</span></code> variable:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [11]: </span><span class="n">ds</span><span class="o">.</span><span class="n">sample_id</span><span class="o">.</span><span class="n">values</span>
<span class="gh">Out[11]: </span><span class="go"></span>
<span class="go">array([&#39;S0&#39;, &#39;S1&#39;, &#39;S2&#39;, &#39;S3&#39;, &#39;S4&#39;, &#39;S5&#39;, &#39;S6&#39;, &#39;S7&#39;, &#39;S8&#39;, &#39;S9&#39;, &#39;S10&#39;,</span>
<span class="go">       &#39;S11&#39;, &#39;S12&#39;, &#39;S13&#39;, &#39;S14&#39;, &#39;S15&#39;, &#39;S16&#39;, &#39;S17&#39;, &#39;S18&#39;, &#39;S19&#39;,</span>
<span class="go">       &#39;S20&#39;, &#39;S21&#39;, &#39;S22&#39;, &#39;S23&#39;, &#39;S24&#39;, &#39;S25&#39;, &#39;S26&#39;, &#39;S27&#39;, &#39;S28&#39;,</span>
<span class="go">       &#39;S29&#39;, &#39;S30&#39;, &#39;S31&#39;, &#39;S32&#39;, &#39;S33&#39;, &#39;S34&#39;, &#39;S35&#39;, &#39;S36&#39;, &#39;S37&#39;,</span>
<span class="go">       &#39;S38&#39;, &#39;S39&#39;, &#39;S40&#39;, &#39;S41&#39;, &#39;S42&#39;, &#39;S43&#39;, &#39;S44&#39;, &#39;S45&#39;, &#39;S46&#39;,</span>
<span class="go">       &#39;S47&#39;, &#39;S48&#39;, &#39;S49&#39;], dtype=&#39;&lt;U3&#39;)</span>
</pre></div>
</div>
</div>
<div class="section" id="subset-the-samples">
<h2><a class="toc-backref" href="#id10">Subset the samples?</a><a class="headerlink" href="#subset-the-samples" title="Permalink to this headline">¶</a></h2>
<p>Call <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.sel.html#xarray.Dataset.sel" title="(in xarray v0.20.2.dev0)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">xarray.Dataset.sel()</span></code></a> and <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.DataArray.isin.html#xarray.DataArray.isin" title="(in xarray v0.20.2.dev0)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">xarray.DataArray.isin()</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [12]: </span><span class="n">ds</span><span class="o">.</span><span class="n">sel</span><span class="p">(</span><span class="n">samples</span><span class="o">=</span><span class="n">ds</span><span class="o">.</span><span class="n">sample_id</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;S30&quot;</span><span class="p">,</span> <span class="s2">&quot;S32&quot;</span><span class="p">]))</span>
<span class="gh">Out[12]: </span><span class="go"></span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:             (variants: 100, alleles: 2, samples: 2, ploidy: 2)</span>
<span class="go">Dimensions without coordinates: variants, alleles, samples, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    variant_contig      (variants) int64 0 0 0 0 0 1 1 ... 21 21 21 22 22 22 22</span>
<span class="go">    variant_position    (variants) int64 0 1 2 3 4 0 1 2 3 ... 3 0 1 2 3 0 1 2 3</span>
<span class="go">    variant_allele      (variants, alleles) |S1 b&#39;T&#39; b&#39;C&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    sample_id           (samples) &lt;U3 &#39;S30&#39; &#39;S32&#39;</span>
<span class="go">    call_genotype       (variants, samples, ploidy) int8 0 -1 0 0 0 ... 1 1 0 0</span>
<span class="go">    call_genotype_mask  (variants, samples, ploidy) bool False True ... False</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
</div>
<div class="section" id="define-a-new-variable-based-on-others">
<h2><a class="toc-backref" href="#id11">Define a new variable based on others?</a><a class="headerlink" href="#define-a-new-variable-based-on-others" title="Permalink to this headline">¶</a></h2>
<p>Use Xarray’s <a class="reference external" href="http://xarray.pydata.org/en/stable/user-guide/data-structures.html#dictionary-like-methods">dictionary like methods</a>, or <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.assign.html#xarray.Dataset.assign" title="(in xarray v0.20.2.dev0)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">xarray.Dataset.assign()</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [13]: </span><span class="n">ds</span><span class="p">[</span><span class="s2">&quot;pos0&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">variant_position</span> <span class="o">-</span> <span class="mi">1</span>

<span class="gp">In [14]: </span><span class="n">ds</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">pos0</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">variant_position</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="c1"># alternative</span>
<span class="gh">Out[14]: </span><span class="go"></span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:             (variants: 100, alleles: 2, samples: 50, ploidy: 2)</span>
<span class="go">Dimensions without coordinates: variants, alleles, samples, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    variant_contig      (variants) int64 0 0 0 0 0 1 1 ... 21 21 21 22 22 22 22</span>
<span class="go">    variant_position    (variants) int64 0 1 2 3 4 0 1 2 3 ... 3 0 1 2 3 0 1 2 3</span>
<span class="go">    variant_allele      (variants, alleles) |S1 b&#39;T&#39; b&#39;C&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    sample_id           (samples) &lt;U3 &#39;S0&#39; &#39;S1&#39; &#39;S2&#39; &#39;S3&#39; ... &#39;S47&#39; &#39;S48&#39; &#39;S49&#39;</span>
<span class="go">    call_genotype       (variants, samples, ploidy) int8 0 0 1 0 1 ... 1 0 1 0 0</span>
<span class="go">    call_genotype_mask  (variants, samples, ploidy) bool False False ... False</span>
<span class="go">    pos0                (variants) int64 -1 0 1 2 3 -1 0 1 ... -1 0 1 2 -1 0 1 2</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
</div>
<div class="section" id="get-summary-stats">
<h2><a class="toc-backref" href="#id12">Get summary stats?</a><a class="headerlink" href="#get-summary-stats" title="Permalink to this headline">¶</a></h2>
<p>Call <a class="reference internal" href="generated/sgkit.sample_stats.html#sgkit.sample_stats" title="sgkit.sample_stats"><code class="xref py py-func docutils literal notranslate"><span class="pre">sample_stats()</span></code></a> or <a class="reference internal" href="generated/sgkit.variant_stats.html#sgkit.variant_stats" title="sgkit.variant_stats"><code class="xref py py-func docutils literal notranslate"><span class="pre">variant_stats()</span></code></a> as appropriate:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [15]: </span><span class="n">sg</span><span class="o">.</span><span class="n">sample_stats</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>
<span class="gh">Out[15]: </span><span class="go"></span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:             (samples: 50, variants: 100, alleles: 2, ploidy: 2)</span>
<span class="go">Dimensions without coordinates: samples, variants, alleles, ploidy</span>
<span class="go">Data variables: (12/13)</span>
<span class="go">    sample_n_called     (samples) int64 81 78 80 77 88 82 ... 76 80 77 90 81 81</span>
<span class="go">    sample_call_rate    (samples) float64 0.81 0.78 0.8 0.77 ... 0.9 0.81 0.81</span>
<span class="go">    sample_n_het        (samples) int64 40 37 40 39 41 33 ... 47 37 40 49 50 37</span>
<span class="go">    sample_n_hom_ref    (samples) int64 20 22 19 21 21 23 ... 9 19 20 17 19 24</span>
<span class="go">    sample_n_hom_alt    (samples) int64 21 19 21 17 26 26 ... 20 24 17 24 12 20</span>
<span class="go">    sample_n_non_ref    (samples) int64 61 56 61 56 67 59 ... 67 61 57 73 62 57</span>
<span class="go">    ...                  ...</span>
<span class="go">    variant_position    (variants) int64 0 1 2 3 4 0 1 2 3 ... 3 0 1 2 3 0 1 2 3</span>
<span class="go">    variant_allele      (variants, alleles) |S1 b&#39;T&#39; b&#39;C&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    sample_id           (samples) &lt;U3 &#39;S0&#39; &#39;S1&#39; &#39;S2&#39; &#39;S3&#39; ... &#39;S47&#39; &#39;S48&#39; &#39;S49&#39;</span>
<span class="go">    call_genotype       (variants, samples, ploidy) int8 0 0 1 0 1 ... 1 0 1 0 0</span>
<span class="go">    call_genotype_mask  (variants, samples, ploidy) bool False False ... False</span>
<span class="go">    pos0                (variants) int64 -1 0 1 2 3 -1 0 1 ... -1 0 1 2 -1 0 1 2</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, ...</span>
<span class="go">    source:   sgkit-unknown</span>

<span class="gp">In [16]: </span><span class="n">sg</span><span class="o">.</span><span class="n">variant_stats</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>
<span class="gh">Out[16]: </span><span class="go"></span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:                   (variants: 100, alleles: 2, samples: 50, ploidy: 2)</span>
<span class="go">Dimensions without coordinates: variants, alleles, samples, ploidy</span>
<span class="go">Data variables: (12/16)</span>
<span class="go">    variant_n_called          (variants) int64 40 42 43 40 41 ... 42 38 43 43 40</span>
<span class="go">    variant_call_rate         (variants) float64 0.8 0.84 0.86 ... 0.86 0.86 0.8</span>
<span class="go">    variant_n_het             (variants) int64 22 30 18 26 29 ... 23 22 26 21 19</span>
<span class="go">    variant_n_hom_ref         (variants) int64 11 3 13 3 6 8 ... 6 10 9 12 8 13</span>
<span class="go">    variant_n_hom_alt         (variants) int64 7 9 12 11 6 17 ... 15 9 7 5 14 8</span>
<span class="go">    variant_n_non_ref         (variants) int64 29 39 30 37 35 ... 32 29 31 35 27</span>
<span class="go">    ...                        ...</span>
<span class="go">    variant_position          (variants) int64 0 1 2 3 4 0 1 2 ... 1 2 3 0 1 2 3</span>
<span class="go">    variant_allele            (variants, alleles) |S1 b&#39;T&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    sample_id                 (samples) &lt;U3 &#39;S0&#39; &#39;S1&#39; &#39;S2&#39; ... &#39;S47&#39; &#39;S48&#39; &#39;S49&#39;</span>
<span class="go">    call_genotype             (variants, samples, ploidy) int8 0 0 1 0 ... 1 0 0</span>
<span class="go">    call_genotype_mask        (variants, samples, ploidy) bool False ... False</span>
<span class="go">    pos0                      (variants) int64 -1 0 1 2 3 -1 0 ... 1 2 -1 0 1 2</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
</div>
<div class="section" id="filter-variants">
<h2><a class="toc-backref" href="#id13">Filter variants?</a><a class="headerlink" href="#filter-variants" title="Permalink to this headline">¶</a></h2>
<p>Call <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.sel.html#xarray.Dataset.sel" title="(in xarray v0.20.2.dev0)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">xarray.Dataset.sel()</span></code></a> on the <code class="docutils literal notranslate"><span class="pre">variants</span></code> dimension:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [17]: </span><span class="n">ds2</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">hardy_weinberg_test</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>

<span class="gp">In [18]: </span><span class="n">ds2</span><span class="o">.</span><span class="n">sel</span><span class="p">(</span><span class="n">variants</span><span class="o">=</span><span class="p">(</span><span class="n">ds2</span><span class="o">.</span><span class="n">variant_hwe_p_value</span> <span class="o">&gt;</span> <span class="mf">1e-2</span><span class="p">))</span>
<span class="gh">Out[18]: </span><span class="go"></span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:              (variants: 99, alleles: 2, samples: 50, ploidy: 2)</span>
<span class="go">Dimensions without coordinates: variants, alleles, samples, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    variant_hwe_p_value  (variants) float64 dask.array&lt;chunksize=(99,), meta=np.ndarray&gt;</span>
<span class="go">    variant_n_het        (variants) int64 22 18 26 29 18 15 ... 23 22 26 21 19</span>
<span class="go">    variant_n_hom_ref    (variants) int64 11 13 3 6 8 13 9 ... 11 6 10 9 12 8 13</span>
<span class="go">    variant_n_hom_alt    (variants) int64 7 12 11 6 17 10 14 ... 9 15 9 7 5 14 8</span>
<span class="go">    variant_n_non_ref    (variants) int64 29 30 37 35 35 25 ... 32 29 31 35 27</span>
<span class="go">    variant_contig       (variants) int64 0 0 0 0 1 1 1 ... 21 21 21 22 22 22 22</span>
<span class="go">    variant_position     (variants) int64 0 2 3 4 0 1 2 3 4 ... 0 1 2 3 0 1 2 3</span>
<span class="go">    variant_allele       (variants, alleles) |S1 b&#39;T&#39; b&#39;C&#39; b&#39;T&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    sample_id            (samples) &lt;U3 &#39;S0&#39; &#39;S1&#39; &#39;S2&#39; &#39;S3&#39; ... &#39;S47&#39; &#39;S48&#39; &#39;S49&#39;</span>
<span class="go">    call_genotype        (variants, samples, ploidy) int8 0 0 1 0 1 ... 0 1 0 0</span>
<span class="go">    call_genotype_mask   (variants, samples, ploidy) bool False False ... False</span>
<span class="go">    pos0                 (variants) int64 -1 1 2 3 -1 0 1 2 ... 0 1 2 -1 0 1 2</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Filtering causes an eager Xarray computation.</p>
</div>
</div>
<div class="section" id="find-which-new-variables-were-added-by-a-method">
<h2><a class="toc-backref" href="#id14">Find which new variables were added by a method?</a><a class="headerlink" href="#find-which-new-variables-were-added-by-a-method" title="Permalink to this headline">¶</a></h2>
<p>Use <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.data_vars.html#xarray.Dataset.data_vars" title="(in xarray v0.20.2.dev0)"><code class="xref py py-attr docutils literal notranslate"><span class="pre">xarray.Dataset.data_vars</span></code></a> to compare the new dataset variables to the old:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [19]: </span><span class="n">ds2</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">sample_stats</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>

<span class="gp">In [20]: </span><span class="nb">set</span><span class="p">(</span><span class="n">ds2</span><span class="o">.</span><span class="n">data_vars</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">data_vars</span><span class="p">)</span>
<span class="gh">Out[20]: </span><span class="go"></span>
<span class="go">{&#39;sample_call_rate&#39;,</span>
<span class="go"> &#39;sample_n_called&#39;,</span>
<span class="go"> &#39;sample_n_het&#39;,</span>
<span class="go"> &#39;sample_n_hom_alt&#39;,</span>
<span class="go"> &#39;sample_n_hom_ref&#39;,</span>
<span class="go"> &#39;sample_n_non_ref&#39;}</span>
</pre></div>
</div>
</div>
<div class="section" id="save-results-to-a-zarr-file">
<h2><a class="toc-backref" href="#id15">Save results to a Zarr file?</a><a class="headerlink" href="#save-results-to-a-zarr-file" title="Permalink to this headline">¶</a></h2>
<p>Call <a class="reference internal" href="generated/sgkit.save_dataset.html#sgkit.save_dataset" title="sgkit.save_dataset"><code class="xref py py-func docutils literal notranslate"><span class="pre">save_dataset()</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [21]: </span><span class="n">sg</span><span class="o">.</span><span class="n">save_dataset</span><span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="s2">&quot;ds.zarr&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Zarr datasets must have equal-sized chunks (except for the final chunk, which may be smaller),
so you may have to <a class="reference external" href="http://xarray.pydata.org/en/stable/generated/xarray.DataArray.chunk.html">rechunk the dataset</a> first.</p>
</div>
</div>
<div class="section" id="load-a-dataset-from-zarr">
<h2><a class="toc-backref" href="#id16">Load a dataset from Zarr?</a><a class="headerlink" href="#load-a-dataset-from-zarr" title="Permalink to this headline">¶</a></h2>
<p>Call <a class="reference internal" href="generated/sgkit.load_dataset.html#sgkit.load_dataset" title="sgkit.load_dataset"><code class="xref py py-func docutils literal notranslate"><span class="pre">load_dataset()</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [22]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="s2">&quot;ds.zarr&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>


              </div>
              
            </div>
        </div>
    </div>
    <div id="main-content" class="row noprint">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="how-do-i">
<span id="id1"></span><h1>How do I …<a class="headerlink" href="#how-do-i" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="contents">
<ul class="simple">
<li><p><a class="reference internal" href="#create-a-test-dataset" id="id2">Create a test dataset?</a></p></li>
<li><p><a class="reference internal" href="#look-at-the-dataset-summary" id="id3">Look at the dataset summary?</a></p></li>
<li><p><a class="reference internal" href="#get-the-values-for-a-variable-in-a-dataset" id="id4">Get the values for a variable in a dataset?</a></p></li>
<li><p><a class="reference internal" href="#find-the-definition-for-a-variable-in-a-dataset" id="id5">Find the definition for a variable in a dataset?</a></p></li>
<li><p><a class="reference internal" href="#look-at-the-genotypes" id="id6">Look at the genotypes?</a></p></li>
<li><p><a class="reference internal" href="#subset-the-variables" id="id7">Subset the variables?</a></p></li>
<li><p><a class="reference internal" href="#subset-to-a-genomic-range" id="id8">Subset to a genomic range?</a></p></li>
<li><p><a class="reference internal" href="#get-the-list-of-samples" id="id9">Get the list of samples?</a></p></li>
<li><p><a class="reference internal" href="#subset-the-samples" id="id10">Subset the samples?</a></p></li>
<li><p><a class="reference internal" href="#define-a-new-variable-based-on-others" id="id11">Define a new variable based on others?</a></p></li>
<li><p><a class="reference internal" href="#get-summary-stats" id="id12">Get summary stats?</a></p></li>
<li><p><a class="reference internal" href="#filter-variants" id="id13">Filter variants?</a></p></li>
<li><p><a class="reference internal" href="#find-which-new-variables-were-added-by-a-method" id="id14">Find which new variables were added by a method?</a></p></li>
<li><p><a class="reference internal" href="#save-results-to-a-zarr-file" id="id15">Save results to a Zarr file?</a></p></li>
<li><p><a class="reference internal" href="#load-a-dataset-from-zarr" id="id16">Load a dataset from Zarr?</a></p></li>
</ul>
</div>
<div class="section" id="create-a-test-dataset">
<h2><a class="toc-backref" href="#id2">Create a test dataset?</a><a class="headerlink" href="#create-a-test-dataset" title="Permalink to this headline">¶</a></h2>
<p>Call <a class="reference internal" href="generated/sgkit.simulate_genotype_call_dataset.html#sgkit.simulate_genotype_call_dataset" title="sgkit.simulate_genotype_call_dataset"><code class="xref py py-func docutils literal notranslate"><span class="pre">simulate_genotype_call_dataset()</span></code></a> to create a test <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v0.20.2.dev0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">xarray.Dataset</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [1]: </span><span class="kn">import</span> <span class="nn">sgkit</span> <span class="k">as</span> <span class="nn">sg</span>

<span class="gp">In [2]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">simulate_genotype_call_dataset</span><span class="p">(</span><span class="n">n_variant</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">n_sample</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">n_contig</span><span class="o">=</span><span class="mi">23</span><span class="p">,</span> <span class="n">missing_pct</span><span class="o">=</span><span class="mf">.1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="look-at-the-dataset-summary">
<h2><a class="toc-backref" href="#id3">Look at the dataset summary?</a><a class="headerlink" href="#look-at-the-dataset-summary" title="Permalink to this headline">¶</a></h2>
<p>Print using the <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v0.20.2.dev0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">xarray.Dataset</span></code></a> <code class="docutils literal notranslate"><span class="pre">repr</span></code>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [3]: </span><span class="n">ds</span>
<span class="gh">Out[3]: </span><span class="go"></span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:             (variants: 100, alleles: 2, samples: 50, ploidy: 2)</span>
<span class="go">Dimensions without coordinates: variants, alleles, samples, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    variant_contig      (variants) int64 0 0 0 0 0 1 1 ... 21 21 21 22 22 22 22</span>
<span class="go">    variant_position    (variants) int64 0 1 2 3 4 0 1 2 3 ... 3 0 1 2 3 0 1 2 3</span>
<span class="go">    variant_allele      (variants, alleles) |S1 b&#39;T&#39; b&#39;C&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    sample_id           (samples) &lt;U3 &#39;S0&#39; &#39;S1&#39; &#39;S2&#39; &#39;S3&#39; ... &#39;S47&#39; &#39;S48&#39; &#39;S49&#39;</span>
<span class="go">    call_genotype       (variants, samples, ploidy) int8 0 0 1 0 1 ... 1 0 1 0 0</span>
<span class="go">    call_genotype_mask  (variants, samples, ploidy) bool False False ... False</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
</div>
<div class="section" id="get-the-values-for-a-variable-in-a-dataset">
<h2><a class="toc-backref" href="#id4">Get the values for a variable in a dataset?</a><a class="headerlink" href="#get-the-values-for-a-variable-in-a-dataset" title="Permalink to this headline">¶</a></h2>
<p>Call <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Variable.values.html#xarray.Variable.values" title="(in xarray v0.20.2.dev0)"><code class="xref py py-attr docutils literal notranslate"><span class="pre">xarray.Variable.values</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [4]: </span><span class="n">ds</span><span class="o">.</span><span class="n">variant_contig</span><span class="o">.</span><span class="n">values</span>
<span class="gh">Out[4]: </span><span class="go"></span>
<span class="go">array([ 0,  0,  0,  0,  0,  1,  1,  1,  1,  1,  2,  2,  2,  2,  2,  3,  3,</span>
<span class="go">        3,  3,  3,  4,  4,  4,  4,  4,  5,  5,  5,  5,  5,  6,  6,  6,  6,</span>
<span class="go">        6,  7,  7,  7,  7,  7,  8,  8,  8,  8,  9,  9,  9,  9, 10, 10, 10,</span>
<span class="go">       10, 11, 11, 11, 11, 12, 12, 12, 12, 13, 13, 13, 13, 14, 14, 14, 14,</span>
<span class="go">       15, 15, 15, 15, 16, 16, 16, 16, 17, 17, 17, 17, 18, 18, 18, 18, 19,</span>
<span class="go">       19, 19, 19, 20, 20, 20, 20, 21, 21, 21, 21, 22, 22, 22, 22])</span>

<span class="gp">In [5]: </span><span class="n">ds</span><span class="p">[</span><span class="s2">&quot;variant_contig&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span> <span class="c1"># equivalent alternative</span>
<span class="gh">Out[5]: </span><span class="go"></span>
<span class="go">array([ 0,  0,  0,  0,  0,  1,  1,  1,  1,  1,  2,  2,  2,  2,  2,  3,  3,</span>
<span class="go">        3,  3,  3,  4,  4,  4,  4,  4,  5,  5,  5,  5,  5,  6,  6,  6,  6,</span>
<span class="go">        6,  7,  7,  7,  7,  7,  8,  8,  8,  8,  9,  9,  9,  9, 10, 10, 10,</span>
<span class="go">       10, 11, 11, 11, 11, 12, 12, 12, 12, 13, 13, 13, 13, 14, 14, 14, 14,</span>
<span class="go">       15, 15, 15, 15, 16, 16, 16, 16, 17, 17, 17, 17, 18, 18, 18, 18, 19,</span>
<span class="go">       19, 19, 19, 20, 20, 20, 20, 21, 21, 21, 21, 22, 22, 22, 22])</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Calling <code class="docutils literal notranslate"><span class="pre">values</span></code> materializes a variable’s data in memory, so is only suitable for small datasets.</p>
</div>
</div>
<div class="section" id="find-the-definition-for-a-variable-in-a-dataset">
<h2><a class="toc-backref" href="#id5">Find the definition for a variable in a dataset?</a><a class="headerlink" href="#find-the-definition-for-a-variable-in-a-dataset" title="Permalink to this headline">¶</a></h2>
<p>Use the <code class="docutils literal notranslate"><span class="pre">comment</span></code> attribute on the variable:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [6]: </span><span class="n">ds</span><span class="o">.</span><span class="n">variant_contig</span><span class="o">.</span><span class="n">comment</span>
<span class="gh">Out[6]: </span><span class="go">&#39;Index corresponding to contig name for each variant. In some less common\nscenarios, this may also be equivalent to the contig names if the data\ngenerating process used contig names that were also integers.&#39;</span>
</pre></div>
</div>
<p>All the variables defined in sgkit are documented on the <a class="reference internal" href="api.html#api-variables"><span class="std std-ref">Variables</span></a> API page.</p>
</div>
<div class="section" id="look-at-the-genotypes">
<h2><a class="toc-backref" href="#id6">Look at the genotypes?</a><a class="headerlink" href="#look-at-the-genotypes" title="Permalink to this headline">¶</a></h2>
<p>Call <a class="reference internal" href="generated/sgkit.display_genotypes.html#sgkit.display_genotypes" title="sgkit.display_genotypes"><code class="xref py py-func docutils literal notranslate"><span class="pre">display_genotypes()</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [7]: </span><span class="n">sg</span><span class="o">.</span><span class="n">display_genotypes</span><span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="n">max_variants</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="gh">Out[7]: </span><span class="go"></span>
<span class="go">samples    S0   S1   S2   S3   S4  ...  S45  S46  S47  S48  S49</span>
<span class="go">variants                           ...                         </span>
<span class="go">0         0/0  1/0  1/0  0/1  1/0  ...  1/1  0/0  1/0  0/0  1/1</span>
<span class="go">1         1/1  1/0  1/.  ./0  1/0  ...  1/1  0/1  1/0  1/1  1/0</span>
<span class="go">2         0/1  1/1  1/1  1/0  1/1  ...  0/0  0/1  0/0  0/0  1/1</span>
<span class="go">3         1/1  0/0  1/1  ./1  0/1  ...  0/1  1/0  0/1  0/.  0/.</span>
<span class="go">4         1/0  0/1  0/1  0/1  0/0  ...  1/0  1/1  0/0  1/.  1/0</span>
<span class="go">...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...</span>
<span class="go">7         1/1  1/1  ./0  1/1  0/1  ...  0/0  0/.  1/0  1/0  0/1</span>
<span class="go">8         1/.  1/0  ./0  0/1  1/0  ...  0/1  1/.  0/0  1/0  0/0</span>
<span class="go">9         0/1  0/0  0/0  0/1  0/0  ...  0/1  0/1  1/0  1/0  0/0</span>
<span class="go">10        1/1  0/0  ./1  1/0  0/0  ...  0/0  0/0  1/1  0/1  1/0</span>
<span class="go">11        1/1  0/.  0/0  0/1  1/.  ...  1/0  0/.  0/1  0/1  0/0</span>

<span class="go">[100 rows x 50 columns]</span>
</pre></div>
</div>
</div>
<div class="section" id="subset-the-variables">
<h2><a class="toc-backref" href="#id7">Subset the variables?</a><a class="headerlink" href="#subset-the-variables" title="Permalink to this headline">¶</a></h2>
<p>Use Xarray’s pandas-like method for <a class="reference external" href="http://xarray.pydata.org/en/latest/user-guide/data-structures.html#transforming-datasets">selecting variables</a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [8]: </span><span class="n">ds</span><span class="p">[[</span><span class="s2">&quot;variant_contig&quot;</span><span class="p">,</span> <span class="s2">&quot;variant_position&quot;</span><span class="p">,</span> <span class="s2">&quot;variant_allele&quot;</span><span class="p">]]</span>
<span class="gh">Out[8]: </span><span class="go"></span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:           (variants: 100, alleles: 2)</span>
<span class="go">Dimensions without coordinates: variants, alleles</span>
<span class="go">Data variables:</span>
<span class="go">    variant_contig    (variants) int64 0 0 0 0 0 1 1 1 ... 21 21 21 22 22 22 22</span>
<span class="go">    variant_position  (variants) int64 0 1 2 3 4 0 1 2 3 4 ... 3 0 1 2 3 0 1 2 3</span>
<span class="go">    variant_allele    (variants, alleles) |S1 b&#39;T&#39; b&#39;C&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
<p>Alternatively, you can <a class="reference external" href="http://xarray.pydata.org/en/latest/generated/xarray.Dataset.drop_vars.html#xarray.Dataset.drop_vars">drop variables</a> that you want to remove:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [9]: </span><span class="n">ds</span><span class="o">.</span><span class="n">drop_vars</span><span class="p">([</span><span class="s2">&quot;variant_contig&quot;</span><span class="p">,</span> <span class="s2">&quot;variant_position&quot;</span><span class="p">,</span> <span class="s2">&quot;variant_allele&quot;</span><span class="p">])</span>
<span class="gh">Out[9]: </span><span class="go"></span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:             (samples: 50, variants: 100, ploidy: 2)</span>
<span class="go">Dimensions without coordinates: samples, variants, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    sample_id           (samples) &lt;U3 &#39;S0&#39; &#39;S1&#39; &#39;S2&#39; &#39;S3&#39; ... &#39;S47&#39; &#39;S48&#39; &#39;S49&#39;</span>
<span class="go">    call_genotype       (variants, samples, ploidy) int8 0 0 1 0 1 ... 1 0 1 0 0</span>
<span class="go">    call_genotype_mask  (variants, samples, ploidy) bool False False ... False</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
</div>
<div class="section" id="subset-to-a-genomic-range">
<h2><a class="toc-backref" href="#id8">Subset to a genomic range?</a><a class="headerlink" href="#subset-to-a-genomic-range" title="Permalink to this headline">¶</a></h2>
<p>Set an index on the dataset, then call <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.sel.html#xarray.Dataset.sel" title="(in xarray v0.20.2.dev0)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">xarray.Dataset.sel()</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [10]: </span><span class="n">ds</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">variants</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;variant_contig&quot;</span><span class="p">,</span> <span class="s2">&quot;variant_position&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">sel</span><span class="p">(</span><span class="n">variants</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">slice</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">)))</span>
<span class="gh">Out[10]: </span><span class="go"></span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:             (variants: 3, alleles: 2, samples: 50, ploidy: 2)</span>
<span class="go">Coordinates:</span>
<span class="go">  * variants            (variants) MultiIndex</span>
<span class="go">  - variant_contig      (variants) int64 0 0 0</span>
<span class="go">  - variant_position    (variants) int64 2 3 4</span>
<span class="go">Dimensions without coordinates: alleles, samples, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    variant_allele      (variants, alleles) |S1 b&#39;T&#39; b&#39;G&#39; b&#39;G&#39; b&#39;G&#39; b&#39;C&#39; b&#39;G&#39;</span>
<span class="go">    sample_id           (samples) &lt;U3 &#39;S0&#39; &#39;S1&#39; &#39;S2&#39; &#39;S3&#39; ... &#39;S47&#39; &#39;S48&#39; &#39;S49&#39;</span>
<span class="go">    call_genotype       (variants, samples, ploidy) int8 0 1 1 1 1 ... 1 -1 1 0</span>
<span class="go">    call_genotype_mask  (variants, samples, ploidy) bool False False ... False</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
<p>An API to make this easier is under discussion. Please add your requirements to <a class="reference external" href="https://github.com/pystatgen/sgkit/pull/658">https://github.com/pystatgen/sgkit/pull/658</a>.</p>
</div>
<div class="section" id="get-the-list-of-samples">
<h2><a class="toc-backref" href="#id9">Get the list of samples?</a><a class="headerlink" href="#get-the-list-of-samples" title="Permalink to this headline">¶</a></h2>
<p>Get the values for the <code class="docutils literal notranslate"><span class="pre">sample_id</span></code> variable:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [11]: </span><span class="n">ds</span><span class="o">.</span><span class="n">sample_id</span><span class="o">.</span><span class="n">values</span>
<span class="gh">Out[11]: </span><span class="go"></span>
<span class="go">array([&#39;S0&#39;, &#39;S1&#39;, &#39;S2&#39;, &#39;S3&#39;, &#39;S4&#39;, &#39;S5&#39;, &#39;S6&#39;, &#39;S7&#39;, &#39;S8&#39;, &#39;S9&#39;, &#39;S10&#39;,</span>
<span class="go">       &#39;S11&#39;, &#39;S12&#39;, &#39;S13&#39;, &#39;S14&#39;, &#39;S15&#39;, &#39;S16&#39;, &#39;S17&#39;, &#39;S18&#39;, &#39;S19&#39;,</span>
<span class="go">       &#39;S20&#39;, &#39;S21&#39;, &#39;S22&#39;, &#39;S23&#39;, &#39;S24&#39;, &#39;S25&#39;, &#39;S26&#39;, &#39;S27&#39;, &#39;S28&#39;,</span>
<span class="go">       &#39;S29&#39;, &#39;S30&#39;, &#39;S31&#39;, &#39;S32&#39;, &#39;S33&#39;, &#39;S34&#39;, &#39;S35&#39;, &#39;S36&#39;, &#39;S37&#39;,</span>
<span class="go">       &#39;S38&#39;, &#39;S39&#39;, &#39;S40&#39;, &#39;S41&#39;, &#39;S42&#39;, &#39;S43&#39;, &#39;S44&#39;, &#39;S45&#39;, &#39;S46&#39;,</span>
<span class="go">       &#39;S47&#39;, &#39;S48&#39;, &#39;S49&#39;], dtype=&#39;&lt;U3&#39;)</span>
</pre></div>
</div>
</div>
<div class="section" id="subset-the-samples">
<h2><a class="toc-backref" href="#id10">Subset the samples?</a><a class="headerlink" href="#subset-the-samples" title="Permalink to this headline">¶</a></h2>
<p>Call <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.sel.html#xarray.Dataset.sel" title="(in xarray v0.20.2.dev0)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">xarray.Dataset.sel()</span></code></a> and <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.DataArray.isin.html#xarray.DataArray.isin" title="(in xarray v0.20.2.dev0)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">xarray.DataArray.isin()</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [12]: </span><span class="n">ds</span><span class="o">.</span><span class="n">sel</span><span class="p">(</span><span class="n">samples</span><span class="o">=</span><span class="n">ds</span><span class="o">.</span><span class="n">sample_id</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;S30&quot;</span><span class="p">,</span> <span class="s2">&quot;S32&quot;</span><span class="p">]))</span>
<span class="gh">Out[12]: </span><span class="go"></span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:             (variants: 100, alleles: 2, samples: 2, ploidy: 2)</span>
<span class="go">Dimensions without coordinates: variants, alleles, samples, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    variant_contig      (variants) int64 0 0 0 0 0 1 1 ... 21 21 21 22 22 22 22</span>
<span class="go">    variant_position    (variants) int64 0 1 2 3 4 0 1 2 3 ... 3 0 1 2 3 0 1 2 3</span>
<span class="go">    variant_allele      (variants, alleles) |S1 b&#39;T&#39; b&#39;C&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    sample_id           (samples) &lt;U3 &#39;S30&#39; &#39;S32&#39;</span>
<span class="go">    call_genotype       (variants, samples, ploidy) int8 0 -1 0 0 0 ... 1 1 0 0</span>
<span class="go">    call_genotype_mask  (variants, samples, ploidy) bool False True ... False</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
</div>
<div class="section" id="define-a-new-variable-based-on-others">
<h2><a class="toc-backref" href="#id11">Define a new variable based on others?</a><a class="headerlink" href="#define-a-new-variable-based-on-others" title="Permalink to this headline">¶</a></h2>
<p>Use Xarray’s <a class="reference external" href="http://xarray.pydata.org/en/stable/user-guide/data-structures.html#dictionary-like-methods">dictionary like methods</a>, or <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.assign.html#xarray.Dataset.assign" title="(in xarray v0.20.2.dev0)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">xarray.Dataset.assign()</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [13]: </span><span class="n">ds</span><span class="p">[</span><span class="s2">&quot;pos0&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">variant_position</span> <span class="o">-</span> <span class="mi">1</span>

<span class="gp">In [14]: </span><span class="n">ds</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">pos0</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">variant_position</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="c1"># alternative</span>
<span class="gh">Out[14]: </span><span class="go"></span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:             (variants: 100, alleles: 2, samples: 50, ploidy: 2)</span>
<span class="go">Dimensions without coordinates: variants, alleles, samples, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    variant_contig      (variants) int64 0 0 0 0 0 1 1 ... 21 21 21 22 22 22 22</span>
<span class="go">    variant_position    (variants) int64 0 1 2 3 4 0 1 2 3 ... 3 0 1 2 3 0 1 2 3</span>
<span class="go">    variant_allele      (variants, alleles) |S1 b&#39;T&#39; b&#39;C&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    sample_id           (samples) &lt;U3 &#39;S0&#39; &#39;S1&#39; &#39;S2&#39; &#39;S3&#39; ... &#39;S47&#39; &#39;S48&#39; &#39;S49&#39;</span>
<span class="go">    call_genotype       (variants, samples, ploidy) int8 0 0 1 0 1 ... 1 0 1 0 0</span>
<span class="go">    call_genotype_mask  (variants, samples, ploidy) bool False False ... False</span>
<span class="go">    pos0                (variants) int64 -1 0 1 2 3 -1 0 1 ... -1 0 1 2 -1 0 1 2</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
</div>
<div class="section" id="get-summary-stats">
<h2><a class="toc-backref" href="#id12">Get summary stats?</a><a class="headerlink" href="#get-summary-stats" title="Permalink to this headline">¶</a></h2>
<p>Call <a class="reference internal" href="generated/sgkit.sample_stats.html#sgkit.sample_stats" title="sgkit.sample_stats"><code class="xref py py-func docutils literal notranslate"><span class="pre">sample_stats()</span></code></a> or <a class="reference internal" href="generated/sgkit.variant_stats.html#sgkit.variant_stats" title="sgkit.variant_stats"><code class="xref py py-func docutils literal notranslate"><span class="pre">variant_stats()</span></code></a> as appropriate:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [15]: </span><span class="n">sg</span><span class="o">.</span><span class="n">sample_stats</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>
<span class="gh">Out[15]: </span><span class="go"></span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:             (samples: 50, variants: 100, alleles: 2, ploidy: 2)</span>
<span class="go">Dimensions without coordinates: samples, variants, alleles, ploidy</span>
<span class="go">Data variables: (12/13)</span>
<span class="go">    sample_n_called     (samples) int64 81 78 80 77 88 82 ... 76 80 77 90 81 81</span>
<span class="go">    sample_call_rate    (samples) float64 0.81 0.78 0.8 0.77 ... 0.9 0.81 0.81</span>
<span class="go">    sample_n_het        (samples) int64 40 37 40 39 41 33 ... 47 37 40 49 50 37</span>
<span class="go">    sample_n_hom_ref    (samples) int64 20 22 19 21 21 23 ... 9 19 20 17 19 24</span>
<span class="go">    sample_n_hom_alt    (samples) int64 21 19 21 17 26 26 ... 20 24 17 24 12 20</span>
<span class="go">    sample_n_non_ref    (samples) int64 61 56 61 56 67 59 ... 67 61 57 73 62 57</span>
<span class="go">    ...                  ...</span>
<span class="go">    variant_position    (variants) int64 0 1 2 3 4 0 1 2 3 ... 3 0 1 2 3 0 1 2 3</span>
<span class="go">    variant_allele      (variants, alleles) |S1 b&#39;T&#39; b&#39;C&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    sample_id           (samples) &lt;U3 &#39;S0&#39; &#39;S1&#39; &#39;S2&#39; &#39;S3&#39; ... &#39;S47&#39; &#39;S48&#39; &#39;S49&#39;</span>
<span class="go">    call_genotype       (variants, samples, ploidy) int8 0 0 1 0 1 ... 1 0 1 0 0</span>
<span class="go">    call_genotype_mask  (variants, samples, ploidy) bool False False ... False</span>
<span class="go">    pos0                (variants) int64 -1 0 1 2 3 -1 0 1 ... -1 0 1 2 -1 0 1 2</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, ...</span>
<span class="go">    source:   sgkit-unknown</span>

<span class="gp">In [16]: </span><span class="n">sg</span><span class="o">.</span><span class="n">variant_stats</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>
<span class="gh">Out[16]: </span><span class="go"></span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:                   (variants: 100, alleles: 2, samples: 50, ploidy: 2)</span>
<span class="go">Dimensions without coordinates: variants, alleles, samples, ploidy</span>
<span class="go">Data variables: (12/16)</span>
<span class="go">    variant_n_called          (variants) int64 40 42 43 40 41 ... 42 38 43 43 40</span>
<span class="go">    variant_call_rate         (variants) float64 0.8 0.84 0.86 ... 0.86 0.86 0.8</span>
<span class="go">    variant_n_het             (variants) int64 22 30 18 26 29 ... 23 22 26 21 19</span>
<span class="go">    variant_n_hom_ref         (variants) int64 11 3 13 3 6 8 ... 6 10 9 12 8 13</span>
<span class="go">    variant_n_hom_alt         (variants) int64 7 9 12 11 6 17 ... 15 9 7 5 14 8</span>
<span class="go">    variant_n_non_ref         (variants) int64 29 39 30 37 35 ... 32 29 31 35 27</span>
<span class="go">    ...                        ...</span>
<span class="go">    variant_position          (variants) int64 0 1 2 3 4 0 1 2 ... 1 2 3 0 1 2 3</span>
<span class="go">    variant_allele            (variants, alleles) |S1 b&#39;T&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    sample_id                 (samples) &lt;U3 &#39;S0&#39; &#39;S1&#39; &#39;S2&#39; ... &#39;S47&#39; &#39;S48&#39; &#39;S49&#39;</span>
<span class="go">    call_genotype             (variants, samples, ploidy) int8 0 0 1 0 ... 1 0 0</span>
<span class="go">    call_genotype_mask        (variants, samples, ploidy) bool False ... False</span>
<span class="go">    pos0                      (variants) int64 -1 0 1 2 3 -1 0 ... 1 2 -1 0 1 2</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
</div>
<div class="section" id="filter-variants">
<h2><a class="toc-backref" href="#id13">Filter variants?</a><a class="headerlink" href="#filter-variants" title="Permalink to this headline">¶</a></h2>
<p>Call <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.sel.html#xarray.Dataset.sel" title="(in xarray v0.20.2.dev0)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">xarray.Dataset.sel()</span></code></a> on the <code class="docutils literal notranslate"><span class="pre">variants</span></code> dimension:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [17]: </span><span class="n">ds2</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">hardy_weinberg_test</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>

<span class="gp">In [18]: </span><span class="n">ds2</span><span class="o">.</span><span class="n">sel</span><span class="p">(</span><span class="n">variants</span><span class="o">=</span><span class="p">(</span><span class="n">ds2</span><span class="o">.</span><span class="n">variant_hwe_p_value</span> <span class="o">&gt;</span> <span class="mf">1e-2</span><span class="p">))</span>
<span class="gh">Out[18]: </span><span class="go"></span>
<span class="go">&lt;xarray.Dataset&gt;</span>
<span class="go">Dimensions:              (variants: 99, alleles: 2, samples: 50, ploidy: 2)</span>
<span class="go">Dimensions without coordinates: variants, alleles, samples, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    variant_hwe_p_value  (variants) float64 dask.array&lt;chunksize=(99,), meta=np.ndarray&gt;</span>
<span class="go">    variant_n_het        (variants) int64 22 18 26 29 18 15 ... 23 22 26 21 19</span>
<span class="go">    variant_n_hom_ref    (variants) int64 11 13 3 6 8 13 9 ... 11 6 10 9 12 8 13</span>
<span class="go">    variant_n_hom_alt    (variants) int64 7 12 11 6 17 10 14 ... 9 15 9 7 5 14 8</span>
<span class="go">    variant_n_non_ref    (variants) int64 29 30 37 35 35 25 ... 32 29 31 35 27</span>
<span class="go">    variant_contig       (variants) int64 0 0 0 0 1 1 1 ... 21 21 21 22 22 22 22</span>
<span class="go">    variant_position     (variants) int64 0 2 3 4 0 1 2 3 4 ... 0 1 2 3 0 1 2 3</span>
<span class="go">    variant_allele       (variants, alleles) |S1 b&#39;T&#39; b&#39;C&#39; b&#39;T&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    sample_id            (samples) &lt;U3 &#39;S0&#39; &#39;S1&#39; &#39;S2&#39; &#39;S3&#39; ... &#39;S47&#39; &#39;S48&#39; &#39;S49&#39;</span>
<span class="go">    call_genotype        (variants, samples, ploidy) int8 0 0 1 0 1 ... 0 1 0 0</span>
<span class="go">    call_genotype_mask   (variants, samples, ploidy) bool False False ... False</span>
<span class="go">    pos0                 (variants) int64 -1 1 2 3 -1 0 1 2 ... 0 1 2 -1 0 1 2</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, ...</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Filtering causes an eager Xarray computation.</p>
</div>
</div>
<div class="section" id="find-which-new-variables-were-added-by-a-method">
<h2><a class="toc-backref" href="#id14">Find which new variables were added by a method?</a><a class="headerlink" href="#find-which-new-variables-were-added-by-a-method" title="Permalink to this headline">¶</a></h2>
<p>Use <a class="reference external" href="https://xarray.pydata.org/en/stable/generated/xarray.Dataset.data_vars.html#xarray.Dataset.data_vars" title="(in xarray v0.20.2.dev0)"><code class="xref py py-attr docutils literal notranslate"><span class="pre">xarray.Dataset.data_vars</span></code></a> to compare the new dataset variables to the old:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [19]: </span><span class="n">ds2</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">sample_stats</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>

<span class="gp">In [20]: </span><span class="nb">set</span><span class="p">(</span><span class="n">ds2</span><span class="o">.</span><span class="n">data_vars</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">data_vars</span><span class="p">)</span>
<span class="gh">Out[20]: </span><span class="go"></span>
<span class="go">{&#39;sample_call_rate&#39;,</span>
<span class="go"> &#39;sample_n_called&#39;,</span>
<span class="go"> &#39;sample_n_het&#39;,</span>
<span class="go"> &#39;sample_n_hom_alt&#39;,</span>
<span class="go"> &#39;sample_n_hom_ref&#39;,</span>
<span class="go"> &#39;sample_n_non_ref&#39;}</span>
</pre></div>
</div>
</div>
<div class="section" id="save-results-to-a-zarr-file">
<h2><a class="toc-backref" href="#id15">Save results to a Zarr file?</a><a class="headerlink" href="#save-results-to-a-zarr-file" title="Permalink to this headline">¶</a></h2>
<p>Call <a class="reference internal" href="generated/sgkit.save_dataset.html#sgkit.save_dataset" title="sgkit.save_dataset"><code class="xref py py-func docutils literal notranslate"><span class="pre">save_dataset()</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [21]: </span><span class="n">sg</span><span class="o">.</span><span class="n">save_dataset</span><span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="s2">&quot;ds.zarr&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Zarr datasets must have equal-sized chunks (except for the final chunk, which may be smaller),
so you may have to <a class="reference external" href="http://xarray.pydata.org/en/stable/generated/xarray.DataArray.chunk.html">rechunk the dataset</a> first.</p>
</div>
</div>
<div class="section" id="load-a-dataset-from-zarr">
<h2><a class="toc-backref" href="#id16">Load a dataset from Zarr?</a><a class="headerlink" href="#load-a-dataset-from-zarr" title="Permalink to this headline">¶</a></h2>
<p>Call <a class="reference internal" href="generated/sgkit.load_dataset.html#sgkit.load_dataset" title="sgkit.load_dataset"><code class="xref py py-func docutils literal notranslate"><span class="pre">load_dataset()</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [22]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="s2">&quot;ds.zarr&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>


              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="generated/sgkit.variables.window_stop_spec.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">sgkit.variables.window_stop_spec</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="contributing.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Contributing to sgkit</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
  <p>
    
      By sgkit developers<br/>
    
        &copy; Copyright 2020, sgkit developers.<br/>
  </p>
</footer>
</main>


      </div>
    </div>
  
  <script src="_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>