

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>User Guide &#8212; sgkit  documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/typehints.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/docsearch.sbt.css" />
    <link rel="stylesheet" type="text/css" href="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.css" />
  
  <!-- So that users can add custom icons -->
  <script src="_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'user_guide';</script>
    <script>
        DOCUMENTATION_OPTIONS.theme_version = '0.16.1';
        DOCUMENTATION_OPTIONS.theme_switcher_json_url = 'https://sgkit-dev.github.io/sgkit/latest/_static/switcher.json';
        DOCUMENTATION_OPTIONS.theme_switcher_version_match = 'latest';
        DOCUMENTATION_OPTIONS.show_version_warning_banner =
            false;
        </script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.js"></script>
    <script defer="defer" src="_static/docsearch.sbt.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Examples" href="examples/index.html" />
    <link rel="prev" title="Getting Started" href="getting_started.html" /> 
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="latest" />   
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/sgkit_trnsprnt.png" class="logo__image only-light" alt="sgkit  documentation - Home"/>
    <img src="_static/sgkit_blue_trnsprnt.png" class="logo__image only-dark pst-js-only" alt="sgkit  documentation - Home"/>
  
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="getting_started.html">
    Getting Started
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="#">
    User Guide
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="examples/index.html">
    Examples
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="api.html">
    API reference
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="how_do_i.html">
    How do I …
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button"
                data-bs-toggle="dropdown" aria-expanded="false"
                aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="contributing.html">
    Contributing to sgkit
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="about.html">
    About
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="news.html">
    News
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="changelog.html">
    Changelog
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">
<div class="version-switcher__container dropdown pst-js-only">
  <button id="pst-version-switcher-button-2"
    type="button"
    class="version-switcher__button btn btn-sm dropdown-toggle"
    data-bs-toggle="dropdown"
    aria-haspopup="listbox"
    aria-controls="pst-version-switcher-list-2"
    aria-label="Version switcher list"
  >
    Choose version  <!-- this text may get changed later by javascript -->
    <span class="caret"></span>
  </button>
  <div id="pst-version-switcher-list-2"
    class="version-switcher__menu dropdown-menu list-group-flush py-0"
    role="listbox" aria-labelledby="pst-version-switcher-button-2">
    <!-- dropdown will be populated by javascript on page load -->
  </div>
</div></div>
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/sgkit-dev/sgkit" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="getting_started.html">
    Getting Started
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="#">
    User Guide
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="examples/index.html">
    Examples
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="api.html">
    API reference
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="how_do_i.html">
    How do I …
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="contributing.html">
    Contributing to sgkit
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="about.html">
    About
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="news.html">
    News
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="changelog.html">
    Changelog
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">
<div class="version-switcher__container dropdown pst-js-only">
  <button id="pst-version-switcher-button-3"
    type="button"
    class="version-switcher__button btn btn-sm dropdown-toggle"
    data-bs-toggle="dropdown"
    aria-haspopup="listbox"
    aria-controls="pst-version-switcher-list-3"
    aria-label="Version switcher list"
  >
    Choose version  <!-- this text may get changed later by javascript -->
    <span class="caret"></span>
  </button>
  <div id="pst-version-switcher-list-3"
    class="version-switcher__menu dropdown-menu list-group-flush py-0"
    role="listbox" aria-labelledby="pst-version-switcher-button-3">
    <!-- dropdown will be populated by javascript on page load -->
  </div>
</div></div>
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/sgkit-dev/sgkit" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"></div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">User Guide</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                   <section id="user-guide">
<h1>User Guide<a class="headerlink" href="#user-guide" title="Permalink to this heading">#</a></h1>
<nav class="contents local" id="table-of-contents">
<p class="topic-title">Table of contents:</p>
<ul class="simple">
<li><p><a class="reference internal" href="#reading-and-writing-genetic-data" id="id5">Reading and writing genetic data</a></p>
<ul>
<li><p><a class="reference internal" href="#installation" id="id6">Installation</a></p></li>
<li><p><a class="reference internal" href="#converting-genetic-data-to-zarr" id="id7">Converting genetic data to Zarr</a></p></li>
<li><p><a class="reference internal" href="#bgen" id="id8">BGEN</a></p></li>
<li><p><a class="reference internal" href="#plink" id="id9">PLINK</a></p></li>
<li><p><a class="reference internal" href="#working-with-cloud-native-data" id="id10">Working with cloud-native data</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#datasets" id="id11">Datasets</a></p>
<ul>
<li><p><a class="reference internal" href="#genetic-variables" id="id12">Genetic variables</a></p></li>
<li><p><a class="reference internal" href="#dataset-merge-behavior" id="id13">Dataset merge behavior</a></p></li>
<li><p><a class="reference internal" href="#interop-with-other-python-libraries" id="id14">Interop with other Python libraries</a></p></li>
<li><p><a class="reference internal" href="#custom-naming-conventions" id="id15">Custom naming conventions</a></p></li>
<li><p><a class="reference internal" href="#adding-custom-data-to-a-dataset" id="id16">Adding custom data to a Dataset</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#methods" id="id17">Methods</a></p>
<ul>
<li><p><a class="reference internal" href="#custom-computations" id="id18">Custom Computations</a></p></li>
<li><p><a class="reference internal" href="#pca" id="id19">PCA</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#deployment" id="id20">Deployment</a></p>
<ul>
<li><p><a class="reference internal" href="#deploying-sgkit-on-a-cluster" id="id21">Deploying sgkit on a cluster</a></p></li>
<li><p><a class="reference internal" href="#using-gpus" id="id22">Using GPUs</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#troubleshooting" id="id23">Troubleshooting</a></p>
<ul>
<li><p><a class="reference internal" href="#monitoring-operations" id="id24">Monitoring operations</a></p></li>
<li><p><a class="reference internal" href="#visualizing-computations" id="id25">Visualizing computations</a></p></li>
<li><p><a class="reference internal" href="#disabling-the-numba-cache" id="id26">Disabling the Numba cache</a></p></li>
</ul>
</li>
</ul>
</nav>
<section id="reading-and-writing-genetic-data">
<span id="id1"></span><h2><a class="toc-backref" href="#id5" role="doc-backlink">Reading and writing genetic data</a><a class="headerlink" href="#reading-and-writing-genetic-data" title="Permalink to this heading">#</a></h2>
<section id="installation">
<span id="id2"></span><h3><a class="toc-backref" href="#id6" role="doc-backlink">Installation</a><a class="headerlink" href="#installation" title="Permalink to this heading">#</a></h3>
<p>Sgkit can read standard genetic file formats, including PLINK and BGEN. For reading VCF,
please use the <a class="reference external" href="https://github.com/sgkit-dev/bio2zarr">bio2zarr</a> package.</p>
<p>If sgkit has been installed using conda, support for reading BGEN and PLINK is included.</p>
<p>If sgkit has been installed using pip, then support for reading these formats is
not included, and requires additional dependencies, which can be installed
as an “extra” feature using pip, as follows.</p>
<p>To install sgkit with BGEN support:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ pip install &#39;sgkit[bgen]&#39;
</pre></div>
</div>
<p>To install sgkit with PLINK support:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ pip install &#39;sgkit[plink]&#39;
</pre></div>
</div>
</section>
<section id="converting-genetic-data-to-zarr">
<h3><a class="toc-backref" href="#id7" role="doc-backlink">Converting genetic data to Zarr</a><a class="headerlink" href="#converting-genetic-data-to-zarr" title="Permalink to this heading">#</a></h3>
<p>There are broadly two ways of loading genetic file format <code class="docutils literal notranslate"><span class="pre">X</span></code> for use in sgkit:</p>
<ol class="arabic simple">
<li><p>a <code class="docutils literal notranslate"><span class="pre">read_X</span></code> function to read the file directly</p></li>
<li><p>a <code class="docutils literal notranslate"><span class="pre">X_to_zarr</span></code> function to convert to Zarr first</p></li>
</ol>
<p>Which one to use depends on the size of the input, as well as the file format (not all file
formats have both options).</p>
<p>Generally speaking, the <code class="docutils literal notranslate"><span class="pre">read_X</span></code> functions are convenience functions that are suitable
for small input files, or where the cost of conversion is small. The <code class="docutils literal notranslate"><span class="pre">X_to_zarr</span></code> functions
are more appropriate for large datasets, since the expensive conversion step need only be
paid once.</p>
<p>After converting a file to Zarr format, it can be loaded with sgkit’s <a class="reference internal" href="generated/sgkit.load_dataset.html#sgkit.load_dataset" title="sgkit.load_dataset"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.load_dataset()</span></code></a>.
This is like Xarray’s <a class="reference external" href="https://docs.xarray.dev/en/stable/generated/xarray.open_zarr.html#xarray.open_zarr" title="(in xarray v2025.3.2.dev0)"><code class="xref py py-func docutils literal notranslate"><span class="pre">xarray.open_zarr()</span></code></a> function, with some useful defaults applied for you.</p>
<p>Similarly, the function <a class="reference internal" href="generated/sgkit.save_dataset.html#sgkit.save_dataset" title="sgkit.save_dataset"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.save_dataset()</span></code></a> can be used to save the dataset in Zarr format.
This can be used to save a newly-loaded dataset, or a dataset with new variables that are the
result of running genetic methods on the data, so it can be loaded again later.</p>
<p>By default, Zarr datasets created by sgkit (and Xarray) have <a class="reference external" href="http://xarray.pydata.org/en/stable/user-guide/io.html#consolidated-metadata">consolidated metadata</a>,
which makes opening the dataset faster.</p>
</section>
<section id="bgen">
<h3><a class="toc-backref" href="#id8" role="doc-backlink">BGEN</a><a class="headerlink" href="#bgen" title="Permalink to this heading">#</a></h3>
<p>The <a class="reference internal" href="generated/sgkit.io.bgen.read_bgen.html#sgkit.io.bgen.read_bgen" title="sgkit.io.bgen.read_bgen"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.bgen.read_bgen()</span></code></a> function loads a single BGEN dataset as Dask
arrays within an <a class="reference external" href="https://docs.xarray.dev/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v2025.3.2.dev0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">xarray.Dataset</span></code></a> from a <code class="docutils literal notranslate"><span class="pre">bgen</span></code> file.</p>
<p>The <a class="reference internal" href="generated/sgkit.io.bgen.bgen_to_zarr.html#sgkit.io.bgen.bgen_to_zarr" title="sgkit.io.bgen.bgen_to_zarr"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.bgen.bgen_to_zarr()</span></code></a> function converts a <code class="docutils literal notranslate"><span class="pre">bgen</span></code> file to Zarr
files stored in sgkit’s Xarray data representation, which can then be opened
as a <a class="reference external" href="https://docs.xarray.dev/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v2025.3.2.dev0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">xarray.Dataset</span></code></a>.</p>
</section>
<section id="plink">
<h3><a class="toc-backref" href="#id9" role="doc-backlink">PLINK</a><a class="headerlink" href="#plink" title="Permalink to this heading">#</a></h3>
<p>The <a class="reference internal" href="generated/sgkit.io.plink.read_plink.html#sgkit.io.plink.read_plink" title="sgkit.io.plink.read_plink"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.plink.read_plink()</span></code></a> function loads a single PLINK dataset as Dask
arrays within an <a class="reference external" href="https://docs.xarray.dev/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v2025.3.2.dev0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">xarray.Dataset</span></code></a> from <code class="docutils literal notranslate"><span class="pre">bed</span></code>, <code class="docutils literal notranslate"><span class="pre">bim</span></code>, and <code class="docutils literal notranslate"><span class="pre">fam</span></code> files.</p>
<p>The <a class="reference internal" href="generated/sgkit.io.plink.write_plink.html#sgkit.io.plink.write_plink" title="sgkit.io.plink.write_plink"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.plink.write_plink()</span></code></a> and <a class="reference internal" href="generated/sgkit.io.plink.zarr_to_plink.html#sgkit.io.plink.zarr_to_plink" title="sgkit.io.plink.zarr_to_plink"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.io.plink.zarr_to_plink()</span></code></a>
functions convert sgkit’s Xarray data representation to PLINK.</p>
</section>
<section id="working-with-cloud-native-data">
<h3><a class="toc-backref" href="#id10" role="doc-backlink">Working with cloud-native data</a><a class="headerlink" href="#working-with-cloud-native-data" title="Permalink to this heading">#</a></h3>
<p>TODO: Show how to read/write Zarr data in cloud storage</p>
</section>
</section>
<section id="datasets">
<h2><a class="toc-backref" href="#id11" role="doc-backlink">Datasets</a><a class="headerlink" href="#datasets" title="Permalink to this heading">#</a></h2>
<section id="genetic-variables">
<span id="id3"></span><h3><a class="toc-backref" href="#id12" role="doc-backlink">Genetic variables</a><a class="headerlink" href="#genetic-variables" title="Permalink to this heading">#</a></h3>
<p>Most <a class="reference internal" href="getting_started.html#genetic-methods"><span class="std std-ref">Genetic methods</span></a> in sgkit operate on a few variables in an Xarray dataset. Variables have
default names, so you can usually just pass in the dataset, but it’s also possible to use different
variable names.</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [1]: </span><span class="kn">import</span><span class="w"> </span><span class="nn">sgkit</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sg</span>

<span class="gp">In [2]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">simulate_genotype_call_dataset</span><span class="p">(</span><span class="n">n_variant</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">n_sample</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">missing_pct</span><span class="o">=</span><span class="mf">.1</span><span class="p">)</span>

<span class="gp">In [3]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">ds</span><span class="p">[[</span><span class="s1">&#39;variant_allele&#39;</span><span class="p">,</span> <span class="s1">&#39;call_genotype&#39;</span><span class="p">]]</span>

<span class="gp">In [4]: </span><span class="n">ds</span>
<span class="gh">Out[4]: </span>
<span class="go">&lt;xarray.Dataset&gt; Size: 10kB</span>
<span class="go">Dimensions:         (variants: 100, alleles: 2, samples: 50, ploidy: 2)</span>
<span class="go">Dimensions without coordinates: variants, alleles, samples, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    variant_allele  (variants, alleles) |S1 200B b&#39;T&#39; b&#39;C&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    call_genotype   (variants, samples, ploidy) int8 10kB 0 0 1 0 1 ... 0 1 0 0</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [&#39;0&#39;]</span>
<span class="go">    source:   sgkit-unknown</span>

<span class="go"># Use the default variable (call_genotype)</span>
<span class="gp">In [5]: </span><span class="n">sg</span><span class="o">.</span><span class="n">count_call_alleles</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span><span class="o">.</span><span class="n">call_allele_count</span>
<span class="gh">Out[5]: </span>
<span class="go">&lt;xarray.DataArray &#39;call_allele_count&#39; (variants: 100, samples: 50, alleles: 2)&gt; Size: 10kB</span>
<span class="go">dask.array&lt;count_alleles, shape=(100, 50, 2), dtype=uint8, chunksize=(100, 50, 2), chunktype=numpy.ndarray&gt;</span>
<span class="go">Dimensions without coordinates: variants, samples, alleles</span>
<span class="go">Attributes:</span>
<span class="go">    comment:  Allele counts. With shape (variants, samples, alleles) and valu...</span>

<span class="go"># Create a copy of the call_genotype variable, and use that to compute counts</span>
<span class="go"># (More realistically, this variable would be created from another computation or input.)</span>
<span class="gp">In [6]: </span><span class="n">ds</span><span class="p">[</span><span class="s2">&quot;my_call_genotype&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ds</span><span class="p">[</span><span class="s2">&quot;call_genotype&quot;</span><span class="p">]</span>

<span class="gp">In [7]: </span><span class="n">sg</span><span class="o">.</span><span class="n">count_call_alleles</span><span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="n">call_genotype</span><span class="o">=</span><span class="s2">&quot;my_call_genotype&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">call_allele_count</span>
<span class="gh">Out[7]: </span>
<span class="go">&lt;xarray.DataArray &#39;call_allele_count&#39; (variants: 100, samples: 50, alleles: 2)&gt; Size: 10kB</span>
<span class="go">dask.array&lt;count_alleles, shape=(100, 50, 2), dtype=uint8, chunksize=(100, 50, 2), chunktype=numpy.ndarray&gt;</span>
<span class="go">Dimensions without coordinates: variants, samples, alleles</span>
<span class="go">Attributes:</span>
<span class="go">    comment:  Allele counts. With shape (variants, samples, alleles) and valu...</span>
</pre></div>
</div>
<p>For a full list of variables and their default names, see <a class="reference internal" href="api.html#api-variables"><span class="std std-ref">Variables</span></a>.</p>
<p>Methods declare the variables that they use directly. If the variable exists in the dataset, then
it will be used for the computation.</p>
<p>If the variable doesn’t exist in the dataset, then it will be computed if the variable name is
the default one. For example, <a class="reference internal" href="generated/sgkit.count_variant_alleles.html#sgkit.count_variant_alleles" title="sgkit.count_variant_alleles"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.count_variant_alleles()</span></code></a> declares
<code class="docutils literal notranslate"><span class="pre">call_allele_count</span></code> as a variable it needs to perform its computation.
If the dataset doesn’t contain <code class="docutils literal notranslate"><span class="pre">call_allele_count</span></code>, then the method will
call <a class="reference internal" href="generated/sgkit.count_call_alleles.html#sgkit.count_call_alleles" title="sgkit.count_call_alleles"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.count_call_alleles()</span></code></a> to populate it, before running its own computation.</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="go"># The following will create call_allele_count and variant_allele_count</span>
<span class="gp">In [8]: </span><span class="n">sg</span><span class="o">.</span><span class="n">count_variant_alleles</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>
<span class="gh">Out[8]: </span>
<span class="go">&lt;xarray.Dataset&gt; Size: 32kB</span>
<span class="go">Dimensions:               (variants: 100, alleles: 2, samples: 50, ploidy: 2)</span>
<span class="go">Dimensions without coordinates: variants, alleles, samples, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    variant_allele_count  (variants, alleles) uint64 2kB dask.array&lt;chunksize=(100, 2), meta=np.ndarray&gt;</span>
<span class="go">    call_allele_count     (variants, samples, alleles) uint8 10kB dask.array&lt;chunksize=(100, 50, 2), meta=np.ndarray&gt;</span>
<span class="go">    variant_allele        (variants, alleles) |S1 200B b&#39;T&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    call_genotype         (variants, samples, ploidy) int8 10kB 0 0 1 ... 1 0 0</span>
<span class="go">    my_call_genotype      (variants, samples, ploidy) int8 10kB 0 0 1 ... 1 0 0</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [&#39;0&#39;]</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
<p>If however a non-default variable name is used and it doesn’t exist in the dataset, then the
intermediate variable is <em>not</em> populated, and an error is raised, since sgkit expects the user
to have created the variable in that case.</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [9]: </span><span class="n">sg</span><span class="o">.</span><span class="n">count_variant_alleles</span><span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="n">call_allele_count</span><span class="o">=</span><span class="s2">&quot;my_call_allele_count&quot;</span><span class="p">)</span>
<span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">ValueError</span><span class="g g-Whitespace">                                </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span><span class="p">[</span><span class="mi">9</span><span class="p">],</span> <span class="n">line</span> <span class="mi">1</span>
<span class="ne">----&gt; </span><span class="mi">1</span> <span class="n">sg</span><span class="o">.</span><span class="n">count_variant_alleles</span><span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="n">call_allele_count</span><span class="o">=</span><span class="s2">&quot;my_call_allele_count&quot;</span><span class="p">)</span>

<span class="nn">File ~/workspace/sgkit/sgkit/stats/aggregation.py:175,</span> in <span class="ni">count_variant_alleles</span><span class="nt">(ds, call_genotype, call_allele_count, using, merge)</span>
<span class="g g-Whitespace">    </span><span class="mi">115</span><span class="w"> </span><span class="sd">&quot;&quot;&quot;Compute allele count from per-sample allele counts, or genotype calls.</span>
<span class="g g-Whitespace">    </span><span class="mi">116</span><span class="sd"> </span>
<span class="g g-Whitespace">    </span><span class="mi">117</span><span class="sd"> Parameters</span>
<span class="sd">   (...)    172        [4, 0]], dtype=uint64)</span>
<span class="g g-Whitespace">    </span><span class="mi">173</span><span class="sd"> &quot;&quot;&quot;</span>
<span class="g g-Whitespace">    </span><span class="mi">174</span> <span class="k">if</span> <span class="n">using</span> <span class="o">==</span> <span class="n">variables</span><span class="o">.</span><span class="n">call_allele_count</span><span class="p">:</span>
<span class="ne">--&gt; </span><span class="mi">175</span>     <span class="n">ds</span> <span class="o">=</span> <span class="n">define_variable_if_absent</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">176</span>         <span class="n">ds</span><span class="p">,</span> <span class="n">variables</span><span class="o">.</span><span class="n">call_allele_count</span><span class="p">,</span> <span class="n">call_allele_count</span><span class="p">,</span> <span class="n">count_call_alleles</span>
<span class="g g-Whitespace">    </span><span class="mi">177</span>     <span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">178</span>     <span class="n">variables</span><span class="o">.</span><span class="n">validate</span><span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="p">{</span><span class="n">call_allele_count</span><span class="p">:</span> <span class="n">variables</span><span class="o">.</span><span class="n">call_allele_count_spec</span><span class="p">})</span>
<span class="g g-Whitespace">    </span><span class="mi">179</span>     <span class="n">AC</span> <span class="o">=</span> <span class="n">ds</span><span class="p">[</span><span class="n">call_allele_count</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">dim</span><span class="o">=</span><span class="s2">&quot;samples&quot;</span><span class="p">)</span>

<span class="nn">File ~/workspace/sgkit/sgkit/utils.py:280,</span> in <span class="ni">define_variable_if_absent</span><span class="nt">(ds, default_variable_name, variable_name, func, **kwargs)</span>
<span class="g g-Whitespace">    </span><span class="mi">278</span>     <span class="k">return</span> <span class="n">ds</span>
<span class="g g-Whitespace">    </span><span class="mi">279</span> <span class="k">if</span> <span class="n">variable_name</span> <span class="o">!=</span> <span class="n">default_variable_name</span><span class="p">:</span>
<span class="ne">--&gt; </span><span class="mi">280</span>     <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">281</span>         <span class="sa">f</span><span class="s2">&quot;Variable &#39;</span><span class="si">{</span><span class="n">variable_name</span><span class="si">}</span><span class="s2">&#39; with non-default name is missing and will not be automatically defined.&quot;</span>
<span class="g g-Whitespace">    </span><span class="mi">282</span>     <span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">283</span> <span class="k">return</span> <span class="n">func</span><span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

<span class="ne">ValueError</span>: Variable &#39;my_call_allele_count&#39; with non-default name is missing and will not be automatically defined.
</pre></div>
</div>
<p>There are also some variables that cannot be automatically defined, such as <code class="docutils literal notranslate"><span class="pre">call_genotype</span></code>,
since it can’t be computed from other data.</p>
</section>
<section id="dataset-merge-behavior">
<span id="dataset-merge"></span><h3><a class="toc-backref" href="#id13" role="doc-backlink">Dataset merge behavior</a><a class="headerlink" href="#dataset-merge-behavior" title="Permalink to this heading">#</a></h3>
<p>Generally, method functions in sgkit compute some new variables based on the
input dataset, then return a new output dataset that consists of the input
dataset plus the new computed variables. The input dataset is unchanged.</p>
<p>This behavior can be controlled using the <code class="docutils literal notranslate"><span class="pre">merge</span></code> parameter. If set to <code class="docutils literal notranslate"><span class="pre">True</span></code>
(the default), then the function will merge the input dataset and the computed
output variables into a single dataset. Output variables will overwrite any
input variables with the same name, and a warning will be issued in this case.
If <code class="docutils literal notranslate"><span class="pre">False</span></code>, the function will return only the computed output variables.</p>
<p>Examples:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [10]: </span><span class="kn">import</span><span class="w"> </span><span class="nn">sgkit</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sg</span>

<span class="gp">In [11]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">simulate_genotype_call_dataset</span><span class="p">(</span><span class="n">n_variant</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">n_sample</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">missing_pct</span><span class="o">=</span><span class="mf">.1</span><span class="p">)</span>

<span class="gp">In [12]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">ds</span><span class="p">[[</span><span class="s1">&#39;variant_allele&#39;</span><span class="p">,</span> <span class="s1">&#39;call_genotype&#39;</span><span class="p">]]</span>

<span class="gp">In [13]: </span><span class="n">ds</span>
<span class="gh">Out[13]: </span>
<span class="go">&lt;xarray.Dataset&gt; Size: 10kB</span>
<span class="go">Dimensions:         (variants: 100, alleles: 2, samples: 50, ploidy: 2)</span>
<span class="go">Dimensions without coordinates: variants, alleles, samples, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    variant_allele  (variants, alleles) |S1 200B b&#39;T&#39; b&#39;C&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    call_genotype   (variants, samples, ploidy) int8 10kB 0 0 1 0 1 ... 0 1 0 0</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [&#39;0&#39;]</span>
<span class="go">    source:   sgkit-unknown</span>

<span class="go"># By default, new variables are merged into a copy of the provided dataset</span>
<span class="gp">In [14]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">count_variant_alleles</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>

<span class="gp">In [15]: </span><span class="n">ds</span>
<span class="gh">Out[15]: </span>
<span class="go">&lt;xarray.Dataset&gt; Size: 22kB</span>
<span class="go">Dimensions:               (variants: 100, alleles: 2, samples: 50, ploidy: 2)</span>
<span class="go">Dimensions without coordinates: variants, alleles, samples, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    variant_allele_count  (variants, alleles) uint64 2kB dask.array&lt;chunksize=(100, 2), meta=np.ndarray&gt;</span>
<span class="go">    call_allele_count     (variants, samples, alleles) uint8 10kB dask.array&lt;chunksize=(100, 50, 2), meta=np.ndarray&gt;</span>
<span class="go">    variant_allele        (variants, alleles) |S1 200B b&#39;T&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    call_genotype         (variants, samples, ploidy) int8 10kB 0 0 1 ... 1 0 0</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [&#39;0&#39;]</span>
<span class="go">    source:   sgkit-unknown</span>

<span class="go"># If an existing variable would be re-defined, a warning is thrown</span>
<span class="gp">In [16]: </span><span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span>

<span class="gp">In [17]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">count_variant_alleles</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>

<span class="gp">In [18]: </span><span class="k">with</span> <span class="n">warnings</span><span class="o">.</span><span class="n">catch_warnings</span><span class="p">(</span><span class="n">record</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="k">as</span> <span class="n">w</span><span class="p">:</span>
<span class="gp">   ....: </span>    <span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">count_variant_alleles</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>
<span class="gp">   ....: </span>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">w</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">category</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">w</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">message</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="gp">   ....: </span>
<span class="go">MergeWarning: The following variables in the input dataset will be replaced in the output: variant_allele_count</span>

<span class="go"># New variables can also be returned in their own dataset</span>
<span class="gp">In [19]: </span><span class="n">sg</span><span class="o">.</span><span class="n">count_variant_alleles</span><span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="n">merge</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gh">Out[19]: </span>
<span class="go">&lt;xarray.Dataset&gt; Size: 2kB</span>
<span class="go">Dimensions:               (variants: 100, alleles: 2)</span>
<span class="go">Dimensions without coordinates: variants, alleles</span>
<span class="go">Data variables:</span>
<span class="go">    variant_allele_count  (variants, alleles) uint64 2kB dask.array&lt;chunksize=(100, 2), meta=np.ndarray&gt;</span>

<span class="go"># This can be useful for merging multiple datasets manually</span>
<span class="gp">In [20]: </span><span class="n">ds</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">sg</span><span class="o">.</span><span class="n">count_variant_alleles</span><span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="n">merge</span><span class="o">=</span><span class="kc">False</span><span class="p">))</span>
<span class="gh">Out[20]: </span>
<span class="go">&lt;xarray.Dataset&gt; Size: 22kB</span>
<span class="go">Dimensions:               (variants: 100, alleles: 2, samples: 50, ploidy: 2)</span>
<span class="go">Dimensions without coordinates: variants, alleles, samples, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    variant_allele_count  (variants, alleles) uint64 2kB dask.array&lt;chunksize=(100, 2), meta=np.ndarray&gt;</span>
<span class="go">    call_allele_count     (variants, samples, alleles) uint8 10kB dask.array&lt;chunksize=(100, 50, 2), meta=np.ndarray&gt;</span>
<span class="go">    variant_allele        (variants, alleles) |S1 200B b&#39;T&#39; b&#39;C&#39; ... b&#39;T&#39; b&#39;A&#39;</span>
<span class="go">    call_genotype         (variants, samples, ploidy) int8 10kB 0 0 1 ... 1 0 0</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [&#39;0&#39;]</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
<p>Merge can be used to rename output variables too.</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [21]: </span><span class="kn">import</span><span class="w"> </span><span class="nn">sgkit</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sg</span>

<span class="gp">In [22]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">simulate_genotype_call_dataset</span><span class="p">(</span><span class="n">n_variant</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">n_sample</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">missing_pct</span><span class="o">=</span><span class="mf">.1</span><span class="p">)</span>

<span class="gp">In [23]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">ds</span><span class="p">[[</span><span class="s1">&#39;variant_allele&#39;</span><span class="p">,</span> <span class="s1">&#39;call_genotype&#39;</span><span class="p">]]</span>

<span class="gp">In [24]: </span><span class="n">ds</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">sg</span><span class="o">.</span><span class="n">count_variant_alleles</span><span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="n">merge</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">variant_allele_count</span><span class="o">=</span><span class="s2">&quot;my_variant_allele_count&quot;</span><span class="p">))</span>
<span class="gh">Out[24]: </span>
<span class="go">&lt;xarray.Dataset&gt; Size: 12kB</span>
<span class="go">Dimensions:                  (variants: 100, alleles: 2, samples: 50, ploidy: 2)</span>
<span class="go">Dimensions without coordinates: variants, alleles, samples, ploidy</span>
<span class="go">Data variables:</span>
<span class="go">    variant_allele           (variants, alleles) |S1 200B b&#39;T&#39; b&#39;C&#39; ... b&#39;A&#39;</span>
<span class="go">    call_genotype            (variants, samples, ploidy) int8 10kB 0 0 1 ... 0 0</span>
<span class="go">    my_variant_allele_count  (variants, alleles) uint64 2kB dask.array&lt;chunksize=(100, 2), meta=np.ndarray&gt;</span>
<span class="go">Attributes:</span>
<span class="go">    contigs:  [&#39;0&#39;]</span>
<span class="go">    source:   sgkit-unknown</span>
</pre></div>
</div>
<p>Note that there is a limitation where intermediate variables (<code class="docutils literal notranslate"><span class="pre">call_allele_count</span></code> in this case)
are not returned if <code class="docutils literal notranslate"><span class="pre">merge=False</span></code>. See <a class="github reference external" href="https://github.com/sgkit-dev/sgkit/issues/405">sgkit-dev/sgkit#405</a>.</p>
</section>
<section id="interop-with-other-python-libraries">
<span id="python-interop"></span><h3><a class="toc-backref" href="#id14" role="doc-backlink">Interop with other Python libraries</a><a class="headerlink" href="#interop-with-other-python-libraries" title="Permalink to this heading">#</a></h3>
<p>It’s usually easier to pass genetic data between Python libraries as simple NumPy arrays,
rather than saving them in files. In sgkit, any data variable can be computed and extracted
as a NumPy array using the <code class="docutils literal notranslate"><span class="pre">.values</span></code> property.</p>
<p>Genetic data is usually stored in a <a class="reference internal" href="generated/sgkit.variables.call_genotype_spec.html#sgkit.variables.call_genotype_spec" title="sgkit.variables.call_genotype_spec"><code class="xref py py-data docutils literal notranslate"><span class="pre">sgkit.variables.call_genotype_spec</span></code></a> array
which has three dimensions (variants, samples and ploidy). This data structure can be
difficult to work with in generic statistical libraries and it is often necessary to
convert genotypes to a single value per call. The <a class="reference internal" href="generated/sgkit.convert_call_to_index.html#sgkit.convert_call_to_index" title="sgkit.convert_call_to_index"><code class="xref py py-func docutils literal notranslate"><span class="pre">sgkit.convert_call_to_index()</span></code></a>
method converts call genotypes into <a class="reference internal" href="generated/sgkit.variables.call_genotype_index_spec.html#sgkit.variables.call_genotype_index_spec" title="sgkit.variables.call_genotype_index_spec"><code class="xref py py-data docutils literal notranslate"><span class="pre">sgkit.variables.call_genotype_index_spec</span></code></a>
which represents each call as a single integer value. For biallelic datasets, this
value is simply the count of the alternate allele. Genotype calls with missing alleles
will be converted to a <code class="docutils literal notranslate"><span class="pre">-1</span></code>.</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [25]: </span><span class="kn">import</span><span class="w"> </span><span class="nn">sgkit</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sg</span>

<span class="go"># Simulate biallelic genotype calls</span>
<span class="gp">In [26]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">simulate_genotype_call_dataset</span><span class="p">(</span><span class="n">n_variant</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">n_sample</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">missing_pct</span><span class="o">=</span><span class="mf">.1</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="gp">In [27]: </span><span class="n">sg</span><span class="o">.</span><span class="n">display_genotypes</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>
<span class="gh">Out[27]: </span>
<span class="go">samples    S0   S1   S2   S3   S4   S5   S6   S7</span>
<span class="go">variants                                        </span>
<span class="go">0         0/0  1/0  1/0  0/1  1/0  0/1  0/0  ./0</span>
<span class="go">1         1/1  0/0  1/0  0/.  1/0  1/1  1/1  1/0</span>
<span class="go">2         1/0  0/0  1/0  1/1  1/1  1/1  0/0  0/.</span>
<span class="go">3         0/.  1/0  1/0  1/.  0/0  1/0  0/1  ./0</span>
<span class="go">4         0/0  0/.  0/1  0/1  0/1  1/0  1/.  0/.</span>
<span class="go">5         0/1  0/0  1/0  0/1  1/0  1/1  0/0  1/0</span>
<span class="go">6         0/0  1/1  1/1  1/0  1/0  ./0  1/0  1/0</span>
<span class="go">7         1/1  1/0  0/0  1/0  1/1  0/0  0/1  0/1</span>
<span class="go">8         1/0  1/1  0/1  1/0  1/0  0/.  1/0  0/1</span>
<span class="go">9         0/1  ./1  1/1  0/.  1/.  1/0  0/1  0/.</span>

<span class="go"># Convert genotype calls into a numpy array of alternate allele counts</span>
<span class="gp">In [28]: </span><span class="n">sg</span><span class="o">.</span><span class="n">convert_call_to_index</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span><span class="o">.</span><span class="n">call_genotype_index</span><span class="o">.</span><span class="n">values</span>
<span class="gh">Out[28]: </span>
<span class="go">array([[ 0,  1,  1,  1,  1,  1,  0, -1],</span>
<span class="go">       [ 2,  0,  1, -1,  1,  2,  2,  1],</span>
<span class="go">       [ 1,  0,  1,  2,  2,  2,  0, -1],</span>
<span class="go">       [-1,  1,  1, -1,  0,  1,  1, -1],</span>
<span class="go">       [ 0, -1,  1,  1,  1,  1, -1, -1],</span>
<span class="go">       [ 1,  0,  1,  1,  1,  2,  0,  1],</span>
<span class="go">       [ 0,  2,  2,  1,  1, -1,  1,  1],</span>
<span class="go">       [ 2,  1,  0,  1,  2,  0,  1,  1],</span>
<span class="go">       [ 1,  2,  1,  1,  1, -1,  1,  1],</span>
<span class="go">       [ 1, -1,  2, -1, -1,  1,  1, -1]])</span>
</pre></div>
</div>
</section>
<section id="custom-naming-conventions">
<h3><a class="toc-backref" href="#id15" role="doc-backlink">Custom naming conventions</a><a class="headerlink" href="#custom-naming-conventions" title="Permalink to this heading">#</a></h3>
<p>TODO: Show to use a custom naming convention via Xarray renaming features.</p>
</section>
<section id="adding-custom-data-to-a-dataset">
<h3><a class="toc-backref" href="#id16" role="doc-backlink">Adding custom data to a Dataset</a><a class="headerlink" href="#adding-custom-data-to-a-dataset" title="Permalink to this heading">#</a></h3>
<p>TODO:  Show how something like sample metadata can be joined to an existing Xarray dataset. Also briefly explain
indexing and uniqueness within Xarray/Pandas, since this is critical for understanding joins.</p>
</section>
</section>
<section id="methods">
<h2><a class="toc-backref" href="#id17" role="doc-backlink">Methods</a><a class="headerlink" href="#methods" title="Permalink to this heading">#</a></h2>
<section id="custom-computations">
<span id="id4"></span><h3><a class="toc-backref" href="#id18" role="doc-backlink">Custom Computations</a><a class="headerlink" href="#custom-computations" title="Permalink to this heading">#</a></h3>
<p>TODO: Finish explaining how Numba works and how users might apply it</p>
<p>Here is an example that demonstrates an alt allele count:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [29]: </span><span class="kn">import</span><span class="w"> </span><span class="nn">numba</span>

<span class="gp">In [30]: </span><span class="kn">import</span><span class="w"> </span><span class="nn">sgkit</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sg</span>

<span class="gp">In [31]: </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>

<span class="gp">In [32]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">simulate_genotype_call_dataset</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">missing_pct</span><span class="o">=</span><span class="mf">.2</span><span class="p">)</span>

<span class="gp">In [33]: </span><span class="k">def</span><span class="w"> </span><span class="nf">alt_allele_count</span><span class="p">(</span><span class="n">gt</span><span class="p">):</span>
<span class="gp">   ....: </span>    <span class="n">out</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">(</span><span class="n">gt</span><span class="o">.</span><span class="n">shape</span><span class="p">[:</span><span class="mi">2</span><span class="p">],</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">)</span>
<span class="gp">   ....: </span>    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">ndindex</span><span class="p">(</span><span class="o">*</span><span class="n">out</span><span class="o">.</span><span class="n">shape</span><span class="p">):</span>
<span class="gp">   ....: </span>        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">gt</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">):</span>
<span class="gp">   ....: </span>            <span class="n">out</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">gt</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
<span class="gp">   ....: </span>    <span class="k">return</span> <span class="n">out</span>
<span class="gp">   ....: </span>

<span class="gp">In [34]: </span><span class="n">numba</span><span class="o">.</span><span class="n">njit</span><span class="p">(</span><span class="n">alt_allele_count</span><span class="p">)(</span><span class="n">ds</span><span class="o">.</span><span class="n">call_genotype</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<span class="gh">Out[34]: </span>
<span class="go">array([[ 0,  1,  1],</span>
<span class="go">       [ 1,  1, -1],</span>
<span class="go">       [-1,  1,  2],</span>
<span class="go">       [ 0, -1, -1],</span>
<span class="go">       [ 1,  2,  2]])</span>
</pre></div>
</div>
</section>
<section id="pca">
<h3><a class="toc-backref" href="#id19" role="doc-backlink">PCA</a><a class="headerlink" href="#pca" title="Permalink to this heading">#</a></h3>
<p>TODO: Describe the upstream tools for PCA (i.e. those in dask-ml/scikit-learn)</p>
</section>
</section>
<section id="deployment">
<h2><a class="toc-backref" href="#id20" role="doc-backlink">Deployment</a><a class="headerlink" href="#deployment" title="Permalink to this heading">#</a></h2>
<section id="deploying-sgkit-on-a-cluster">
<h3><a class="toc-backref" href="#id21" role="doc-backlink">Deploying sgkit on a cluster</a><a class="headerlink" href="#deploying-sgkit-on-a-cluster" title="Permalink to this heading">#</a></h3>
<p>For parallelism sgkit uses <a class="reference external" href="https://dask.org/">Dask</a> to distribute work across workers.
These workers can either be local processes or remote processes running on a cluster.
By default, dask creates a local cluster with one worker thread per CPU core on the machine.
This is useful for testing and development, but for larger datasets it is often necessary
to run on a cluster. There are several options for this including:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference external" href="https://kubernetes.dask.org/en/latest/">Dask Kubernetes</a></p></li>
<li><p><a class="reference external" href="https://jobqueue.dask.org/en/latest/">Dask Jobqueue (PBS, Slurm, MOAB, SGE, LSF, and HTCondor.)</a></p></li>
<li><p><a class="reference external" href="https://cloudprovider.dask.org/en/latest/">Dask Cloud Provider (AWS, GCP, Azure, etc.)</a></p></li>
<li><p><a class="reference external" href="https://mpi.dask.org/en/latest/">Dask MPI</a></p></li>
<li><p><a class="reference external" href="https://coiled.io/">Coiled</a></p></li>
</ul>
</div></blockquote>
<p>Research institutions often use a job scheduler like Slurm, LFS or PBS to manage compute resources,
so here is a worked example of using Dask Jobqueue to run sgkit on an LSF cluster. The
process is similar for other schedulers, see the
<a class="reference external" href="https://jobqueue.dask.org/en/latest/">Dask Jobqueue documentation</a></p>
<p>The first step is to create a dask scheduler that sgkit connects to. It is often desirable
to have this running in a separate process, in a python REPL or notebook, so that its scaling
can be adjusted on the fly. This needs to be done on a login node or other machine that has
job submission privileges.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">dask_jobqueue</span><span class="w"> </span><span class="kn">import</span> <span class="n">LSFCluster</span>
<span class="n">cluster</span> <span class="o">=</span> <span class="n">LSFCluster</span><span class="p">(</span>
    <span class="n">cores</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>           <span class="c1"># Number of cores per-worker</span>
    <span class="n">memory</span><span class="o">=</span><span class="s2">&quot;16GB&quot;</span><span class="p">,</span>     <span class="c1"># Amount of memory per-worker</span>
    <span class="n">project</span><span class="o">=</span><span class="s2">&quot;myproject&quot;</span><span class="p">,</span>
    <span class="n">queue</span><span class="o">=</span><span class="s2">&quot;myqueue&quot;</span><span class="p">,</span>   <span class="c1"># LSF queue</span>
    <span class="n">walltime</span><span class="o">=</span><span class="s2">&quot;04:00&quot;</span><span class="p">,</span>  <span class="c1"># Set this to a reasonable value for your data size</span>
    <span class="n">use_stdin</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>    <span class="c1"># This is often required for dask to work with LSF</span>
    <span class="n">lsf_units</span><span class="o">=</span><span class="s2">&quot;MB&quot;</span><span class="p">,</span>    <span class="c1"># This may very depending on your clusters config</span>
    <span class="c1"># Any bash commands needed to setup the environment</span>
    <span class="n">job_script_prolouge</span><span class="o">=</span><span class="s2">&quot;module load example/3.8.5&quot;</span><span class="p">,</span>
    <span class="c1"># This last line is useful to gracefully rotate out workers as they reach</span>
    <span class="c1"># the maximum wallclock time for the given queue. This allows a long-running</span>
    <span class="c1"># cluster to run on queues with shorter wallclock limits, but that likely</span>
    <span class="c1"># have higher priority.</span>
    <span class="n">worker_extra_args</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;--lifetime&quot;</span><span class="p">,</span> <span class="s2">&quot;350m&quot;</span><span class="p">,</span> <span class="s2">&quot;--lifetime-stagger&quot;</span><span class="p">,</span> <span class="s2">&quot;5m&quot;</span><span class="p">],</span>
<span class="p">)</span>
</pre></div>
</div>
<p>Now that the cluster is created we can view its dashboard at the address from
<code class="docutils literal notranslate"><span class="pre">cluster.dashboard_link</span></code>, at this point it will have zero workers.</p>
<p>To launch workers we can request a fixed amount with:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cluster</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
<p>At which point the worker jobs will be scheduled and once they are running they will connect
to the dask scheduler and be visible in the dashboard.</p>
<p>The other option is to let dask adaptively scale the number of workers based on the amount
of task that are queued up. This can prevent idle workers when there are no tasks to run,
but there may be a short delay when new tasks are submitted while the workers are being
scheduled. To enable this we can use:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cluster</span><span class="o">.</span><span class="n">adapt</span><span class="p">(</span><span class="n">minimum</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">maximum</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">wait_count</span><span class="o">=</span><span class="mi">120</span><span class="p">)</span>
</pre></div>
</div>
<p>The minimum and maximum are optional, but it is recommended to set them to reasonable values
to prevent the cluster launching thousands of jobs. The wait_count is the number of update
cycles that dask will wait before scaling down the cluster when there are no tasks to run.
By default the update interval is 1s so the above command will wait 120s before scaling down.</p>
<p>Now that the cluster is running we can connect to it from our sgkit code. The scheduler
address is available from <code class="docutils literal notranslate"><span class="pre">cluster.scheduler_address</span></code>. Then in your sgkit code pass this
value to the dask <code class="docutils literal notranslate"><span class="pre">Client</span></code> constructor:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">sgkit</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sg</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">dask.distributed</span><span class="w"> </span><span class="kn">import</span> <span class="n">Client</span>
<span class="n">client</span> <span class="o">=</span> <span class="n">Client</span><span class="p">(</span><span class="s2">&quot;http://scheduler-address:8786&quot;</span><span class="p">)</span>
<span class="c1"># Now run sgkit code as normal</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">sgkit</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="s2">&quot;mydata.zarr&quot;</span><span class="p">)</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">sgkit</span><span class="o">.</span><span class="n">variant_stats</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="using-gpus">
<h3><a class="toc-backref" href="#id22" role="doc-backlink">Using GPUs</a><a class="headerlink" href="#using-gpus" title="Permalink to this heading">#</a></h3>
<p>TODO: Show CuPy examples</p>
</section>
</section>
<section id="troubleshooting">
<h2><a class="toc-backref" href="#id23" role="doc-backlink">Troubleshooting</a><a class="headerlink" href="#troubleshooting" title="Permalink to this heading">#</a></h2>
<section id="monitoring-operations">
<h3><a class="toc-backref" href="#id24" role="doc-backlink">Monitoring operations</a><a class="headerlink" href="#monitoring-operations" title="Permalink to this heading">#</a></h3>
<p>The simplest way to monitor operations when running sgkit on a single host is to use <a class="reference external" href="https://docs.dask.org/en/latest/diagnostics-local.html">Dask local diagnostics</a>.</p>
<p>As an example, this code shows how to track the progress of a single sgkit function:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [35]: </span><span class="kn">import</span><span class="w"> </span><span class="nn">sgkit</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sg</span>

<span class="gp">In [36]: </span><span class="kn">from</span><span class="w"> </span><span class="nn">dask.diagnostics</span><span class="w"> </span><span class="kn">import</span> <span class="n">ProgressBar</span>

<span class="gp">In [37]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">simulate_genotype_call_dataset</span><span class="p">(</span><span class="n">n_variant</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">n_sample</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">missing_pct</span><span class="o">=</span><span class="mf">.1</span><span class="p">)</span>

<span class="gp">In [38]: </span><span class="k">with</span> <span class="n">ProgressBar</span><span class="p">():</span>
<span class="gp">   ....: </span>    <span class="n">ac</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">count_variant_alleles</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span><span class="o">.</span><span class="n">variant_allele_count</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>
<span class="gp">   ....: </span>

<span class="go">[                                        ] | 0% Completed | 85.75 us</span>
<span class="go">[########################################] | 100% Completed | 105.18 ms</span>

<span class="gp">In [39]: </span><span class="n">ac</span><span class="p">[:</span><span class="mi">5</span><span class="p">]</span>
<span class="gh">Out[39]: </span>
<span class="go">&lt;xarray.DataArray &#39;variant_allele_count&#39; (variants: 5, alleles: 2)&gt; Size: 80B</span>
<span class="go">array([[50, 40],</span>
<span class="go">       [40, 52],</span>
<span class="go">       [48, 45],</span>
<span class="go">       [37, 51],</span>
<span class="go">       [45, 46]], dtype=uint64)</span>
<span class="go">Dimensions without coordinates: variants, alleles</span>
<span class="go">Attributes:</span>
<span class="go">    comment:  Variant allele counts. With shape (variants, alleles) and value...</span>
</pre></div>
</div>
<p>Monitoring resource utilization with <a class="reference external" href="https://docs.dask.org/en/latest/diagnostics-local.html#resourceprofiler">ResourceProfiler</a>
and profiling task streams with <a class="reference external" href="https://docs.dask.org/en/latest/diagnostics-local.html#profiler">Profiler</a> are other
commonly used local diagnostics.</p>
<p>For similar monitoring in a distributed cluster, see <a class="reference external" href="https://docs.dask.org/en/latest/diagnostics-distributed.html">Dask distributed diagnostics</a>.</p>
</section>
<section id="visualizing-computations">
<h3><a class="toc-backref" href="#id25" role="doc-backlink">Visualizing computations</a><a class="headerlink" href="#visualizing-computations" title="Permalink to this heading">#</a></h3>
<p>Dask allows you to <a class="reference external" href="https://docs.dask.org/en/latest/graphviz.html">visualize the task graph</a> of a computation
before running it, which can be handy when trying to understand where the bottlenecks are.</p>
<p>In most cases the number of tasks is too large to visualize, so it’s useful to restrict
the graph just a few chunks, as shown in this example.</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [40]: </span><span class="kn">import</span><span class="w"> </span><span class="nn">sgkit</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sg</span>

<span class="gp">In [41]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">simulate_genotype_call_dataset</span><span class="p">(</span><span class="n">n_variant</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">n_sample</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">missing_pct</span><span class="o">=</span><span class="mf">.1</span><span class="p">)</span>

<span class="go"># Rechunk to illustrate multiple tasks</span>
<span class="gp">In [42]: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">chunk</span><span class="p">({</span><span class="s2">&quot;variants&quot;</span><span class="p">:</span> <span class="mi">25</span><span class="p">,</span> <span class="s2">&quot;samples&quot;</span><span class="p">:</span> <span class="mi">25</span><span class="p">})</span>

<span class="gp">In [43]: </span><span class="n">counts</span> <span class="o">=</span> <span class="n">sg</span><span class="o">.</span><span class="n">count_call_alleles</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span><span class="o">.</span><span class="n">call_allele_count</span><span class="o">.</span><span class="n">data</span>

<span class="go"># Restrict to first 3 chunks in variants dimension</span>
<span class="gp">In [44]: </span><span class="n">counts</span> <span class="o">=</span> <span class="n">counts</span><span class="p">[:</span><span class="mi">3</span><span class="o">*</span><span class="n">counts</span><span class="o">.</span><span class="n">chunksize</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="o">...</span><span class="p">]</span>

<span class="gp">In [45]: </span><span class="n">counts</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">optimize_graph</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gh">Out[45]: </span><span class="go">&lt;IPython.core.display.Image object&gt;</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/mydask.png"><img alt="_images/mydask.png" class="align-center" src="_images/mydask.png" style="width: 600px;" /></a>
<p>By passing keyword arguments to <code class="docutils literal notranslate"><span class="pre">visualize</span></code> we can see the order tasks will run in:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="go"># Graph where colors indicate task ordering</span>
<span class="gp">In [46]: </span><span class="n">counts</span><span class="o">.</span><span class="n">visualize</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;order&quot;</span><span class="p">,</span> <span class="n">optimize_graph</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;order&quot;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;autumn&quot;</span><span class="p">,</span> <span class="n">node_attr</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;penwidth&quot;</span><span class="p">:</span> <span class="s2">&quot;4&quot;</span><span class="p">})</span>
<span class="gh">Out[46]: </span><span class="go">&lt;IPython.core.display.Image object&gt;</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/order.png"><img alt="_images/order.png" class="align-center" src="_images/order.png" style="width: 600px;" /></a>
<p>Task order number is shown in circular boxes, colored from red to yellow.</p>
</section>
<section id="disabling-the-numba-cache">
<h3><a class="toc-backref" href="#id26" role="doc-backlink">Disabling the Numba cache</a><a class="headerlink" href="#disabling-the-numba-cache" title="Permalink to this heading">#</a></h3>
<p>Internally, sgkit uses the <a class="reference external" href="https://numba.pydata.org/">Numba JIT compiler</a> to accelerate some methods.
These methods will be compiled the first time that they are used following a new installation.
The compiled methods are automatically cached so that recompilation is not required in future sessions.
However, this may occasionally cause issues with some setups.</p>
<p>Caching of compiled sgkit methods can be disabled by setting the environment variable <code class="docutils literal notranslate"><span class="pre">SGKIT_DISABLE_NUMBA_CACHE</span></code> to <code class="docutils literal notranslate"><span class="pre">1</span></code>.
This variable can also be set from within a python session <em>before</em> loading sgkit.</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [47]: </span><span class="kn">import</span><span class="w"> </span><span class="nn">os</span>

<span class="gp">In [48]: </span><span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s1">&#39;SGKIT_DISABLE_NUMBA_CACHE&#39;</span><span class="p">]</span><span class="o">=</span><span class="s1">&#39;1&#39;</span>

<span class="gp">In [49]: </span><span class="kn">import</span><span class="w"> </span><span class="nn">sgkit</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sg</span>
</pre></div>
</div>
<p>With caching disabled, these methods will be compiled the first time that they are called during each session.
Refer to the <a class="reference external" href="https://numba.readthedocs.io/en/stable/developer/caching.html">Numba notes on caching</a> for more information.</p>
</section>
</section>
</section>

<div class="section ablog__blog_comments">
   
</div>

                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="getting_started.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Getting Started</p>
      </div>
    </a>
    <a class="right-next"
       href="examples/index.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Examples</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#reading-and-writing-genetic-data">Reading and writing genetic data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#installation">Installation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#converting-genetic-data-to-zarr">Converting genetic data to Zarr</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#bgen">BGEN</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#plink">PLINK</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#working-with-cloud-native-data">Working with cloud-native data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#datasets">Datasets</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#genetic-variables">Genetic variables</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#dataset-merge-behavior">Dataset merge behavior</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#interop-with-other-python-libraries">Interop with other Python libraries</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#custom-naming-conventions">Custom naming conventions</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#adding-custom-data-to-a-dataset">Adding custom data to a Dataset</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#methods">Methods</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#custom-computations">Custom Computations</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pca">PCA</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#deployment">Deployment</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#deploying-sgkit-on-a-cluster">Deploying sgkit on a cluster</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#using-gpus">Using GPUs</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#troubleshooting">Troubleshooting</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#monitoring-operations">Monitoring operations</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualizing-computations">Visualizing computations</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#disabling-the-numba-cache">Disabling the Numba cache</a></li>
</ul>
</li>
</ul>
  </nav></div>

  <div class="sidebar-secondary-item">
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/user_guide.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2020, sgkit developers.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 6.2.1.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>