

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>sgkit.simulate_genotype_call_dataset &#8212; sgkit  documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/typehints.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/docsearch.sbt.css" />
    <link rel="stylesheet" type="text/css" href="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.css" />
  
  <!-- So that users can add custom icons -->
  <script src="../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'generated/sgkit.simulate_genotype_call_dataset';</script>
    <script>
        DOCUMENTATION_OPTIONS.theme_version = '0.16.1';
        DOCUMENTATION_OPTIONS.theme_switcher_json_url = 'https://sgkit-dev.github.io/sgkit/latest/_static/switcher.json';
        DOCUMENTATION_OPTIONS.theme_switcher_version_match = 'latest';
        DOCUMENTATION_OPTIONS.show_version_warning_banner =
            false;
        </script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.js"></script>
    <script defer="defer" src="../_static/docsearch.sbt.js"></script>
    <link rel="author" title="About these documents" href="../about.html" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="sgkit.window_by_genome" href="sgkit.window_by_genome.html" />
    <link rel="prev" title="sgkit.simulate_genedrop" href="sgkit.simulate_genedrop.html" /> 
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="latest" />   
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/sgkit_trnsprnt.png" class="logo__image only-light" alt="sgkit  documentation - Home"/>
    <img src="../_static/sgkit_blue_trnsprnt.png" class="logo__image only-dark pst-js-only" alt="sgkit  documentation - Home"/>
  
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../getting_started.html">
    Getting Started
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../user_guide.html">
    User Guide
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../examples/index.html">
    Examples
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../api.html">
    API reference
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../how_do_i.html">
    How do I â€¦
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button"
                data-bs-toggle="dropdown" aria-expanded="false"
                aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../contributing.html">
    Contributing to sgkit
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../about.html">
    About
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../news.html">
    News
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../changelog.html">
    Changelog
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">
<div class="version-switcher__container dropdown pst-js-only">
  <button id="pst-version-switcher-button-2"
    type="button"
    class="version-switcher__button btn btn-sm dropdown-toggle"
    data-bs-toggle="dropdown"
    aria-haspopup="listbox"
    aria-controls="pst-version-switcher-list-2"
    aria-label="Version switcher list"
  >
    Choose version  <!-- this text may get changed later by javascript -->
    <span class="caret"></span>
  </button>
  <div id="pst-version-switcher-list-2"
    class="version-switcher__menu dropdown-menu list-group-flush py-0"
    role="listbox" aria-labelledby="pst-version-switcher-button-2">
    <!-- dropdown will be populated by javascript on page load -->
  </div>
</div></div>
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/sgkit-dev/sgkit" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../getting_started.html">
    Getting Started
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../user_guide.html">
    User Guide
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../examples/index.html">
    Examples
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../api.html">
    API reference
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../how_do_i.html">
    How do I â€¦
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../contributing.html">
    Contributing to sgkit
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../about.html">
    About
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../news.html">
    News
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../changelog.html">
    Changelog
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">
<div class="version-switcher__container dropdown pst-js-only">
  <button id="pst-version-switcher-button-3"
    type="button"
    class="version-switcher__button btn btn-sm dropdown-toggle"
    data-bs-toggle="dropdown"
    aria-haspopup="listbox"
    aria-controls="pst-version-switcher-list-3"
    aria-label="Version switcher list"
  >
    Choose version  <!-- this text may get changed later by javascript -->
    <span class="caret"></span>
  </button>
  <div id="pst-version-switcher-list-3"
    class="version-switcher__menu dropdown-menu list-group-flush py-0"
    role="listbox" aria-labelledby="pst-version-switcher-button-3">
    <!-- dropdown will be populated by javascript on page load -->
  </div>
</div></div>
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/sgkit-dev/sgkit" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="sgkit.io.bgen.bgen_to_zarr.html">sgkit.io.bgen.bgen_to_zarr</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.io.bgen.read_bgen.html">sgkit.io.bgen.read_bgen</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.io.bgen.rechunk_bgen.html">sgkit.io.bgen.rechunk_bgen</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="sgkit.io.plink.plink_to_zarr.html">sgkit.io.plink.plink_to_zarr</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.io.plink.read_plink.html">sgkit.io.plink.read_plink</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.io.plink.write_plink.html">sgkit.io.plink.write_plink</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.io.plink.zarr_to_plink.html">sgkit.io.plink.zarr_to_plink</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="sgkit.load_dataset.html">sgkit.load_dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.save_dataset.html">sgkit.save_dataset</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="sgkit.call_allele_frequencies.html">sgkit.call_allele_frequencies</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.cohort_allele_frequencies.html">sgkit.cohort_allele_frequencies</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.count_call_alleles.html">sgkit.count_call_alleles</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.count_cohort_alleles.html">sgkit.count_cohort_alleles</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.count_variant_alleles.html">sgkit.count_variant_alleles</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.count_variant_genotypes.html">sgkit.count_variant_genotypes</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variant_stats.html">sgkit.variant_stats</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.sample_stats.html">sgkit.sample_stats</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="sgkit.divergence.html">sgkit.divergence</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.diversity.html">sgkit.diversity</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.Fst.html">sgkit.Fst</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.Garud_H.html">sgkit.Garud_H</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.individual_heterozygosity.html">sgkit.individual_heterozygosity</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.observed_heterozygosity.html">sgkit.observed_heterozygosity</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.pbs.html">sgkit.pbs</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.Tajimas_D.html">sgkit.Tajimas_D</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="sgkit.Weir_Goudet_beta.html">sgkit.Weir_Goudet_beta</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.genomic_relationship.html">sgkit.genomic_relationship</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.pc_relate.html">sgkit.pc_relate</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.pedigree_contribution.html">sgkit.pedigree_contribution</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.pedigree_inbreeding.html">sgkit.pedigree_inbreeding</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.pedigree_inverse_kinship.html">sgkit.pedigree_inverse_kinship</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.pedigree_kinship.html">sgkit.pedigree_kinship</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.hybrid_inverse_relationship.html">sgkit.hybrid_inverse_relationship</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.hybrid_relationship.html">sgkit.hybrid_relationship</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.identity_by_state.html">sgkit.identity_by_state</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="sgkit.ld_matrix.html">sgkit.ld_matrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.ld_prune.html">sgkit.ld_prune</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="sgkit.gwas_linear_regression.html">sgkit.gwas_linear_regression</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.regenie.html">sgkit.regenie</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.hardy_weinberg_test.html">sgkit.hardy_weinberg_test</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.genee.html">sgkit.genee</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="sgkit.maximal_independent_set.html">sgkit.maximal_independent_set</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.pairwise_distance.html">sgkit.pairwise_distance</a></li>
</ul>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="sgkit.convert_call_to_index.html">sgkit.convert_call_to_index</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.convert_probability_to_call.html">sgkit.convert_probability_to_call</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.display_genotypes.html">sgkit.display_genotypes</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.display_pedigree.html">sgkit.display_pedigree</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.filter_partial_calls.html">sgkit.filter_partial_calls</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.infer_call_ploidy.html">sgkit.infer_call_ploidy</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.infer_sample_ploidy.html">sgkit.infer_sample_ploidy</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.infer_variant_ploidy.html">sgkit.infer_variant_ploidy</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.invert_relationship_matrix.html">sgkit.invert_relationship_matrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.parent_indices.html">sgkit.parent_indices</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.pedigree_sel.html">sgkit.pedigree_sel</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.simulate_genedrop.html">sgkit.simulate_genedrop</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">sgkit.simulate_genotype_call_dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.window_by_genome.html">sgkit.window_by_genome</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.window_by_interval.html">sgkit.window_by_interval</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.window_by_position.html">sgkit.window_by_position</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.window_by_variant.html">sgkit.window_by_variant</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.call_allele_count_spec.html">sgkit.variables.call_allele_count_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.call_allele_frequency_spec.html">sgkit.variables.call_allele_frequency_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.call_dosage_spec.html">sgkit.variables.call_dosage_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.call_dosage_mask_spec.html">sgkit.variables.call_dosage_mask_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.call_genotype_complete_spec.html">sgkit.variables.call_genotype_complete_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.call_genotype_complete_mask_spec.html">sgkit.variables.call_genotype_complete_mask_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.call_genotype_spec.html">sgkit.variables.call_genotype_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.call_genotype_mask_spec.html">sgkit.variables.call_genotype_mask_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.call_genotype_fill_spec.html">sgkit.variables.call_genotype_fill_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.call_genotype_phased_spec.html">sgkit.variables.call_genotype_phased_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.call_genotype_probability_spec.html">sgkit.variables.call_genotype_probability_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.call_genotype_probability_mask_spec.html">sgkit.variables.call_genotype_probability_mask_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.call_genotype_index_spec.html">sgkit.variables.call_genotype_index_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.call_genotype_index_mask_spec.html">sgkit.variables.call_genotype_index_mask_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.call_heterozygosity_spec.html">sgkit.variables.call_heterozygosity_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.call_ploidy_spec.html">sgkit.variables.call_ploidy_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.cohort_allele_count_spec.html">sgkit.variables.cohort_allele_count_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.cohort_allele_frequency_spec.html">sgkit.variables.cohort_allele_frequency_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.covariates_spec.html">sgkit.variables.covariates_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.interval_contig_name_spec.html">sgkit.variables.interval_contig_name_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.interval_start_spec.html">sgkit.variables.interval_start_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.interval_stop_spec.html">sgkit.variables.interval_stop_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.ld_prune_index_to_drop_spec.html">sgkit.variables.ld_prune_index_to_drop_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.regenie_base_prediction_spec.html">sgkit.variables.regenie_base_prediction_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.regenie_loco_prediction_spec.html">sgkit.variables.regenie_loco_prediction_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.regenie_meta_prediction_spec.html">sgkit.variables.regenie_meta_prediction_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.parent_spec.html">sgkit.variables.parent_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.parent_id_spec.html">sgkit.variables.parent_id_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.pc_relate_phi_spec.html">sgkit.variables.pc_relate_phi_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.sample_call_rate_spec.html">sgkit.variables.sample_call_rate_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.sample_cohort_spec.html">sgkit.variables.sample_cohort_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.sample_id_spec.html">sgkit.variables.sample_id_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.sample_n_called_spec.html">sgkit.variables.sample_n_called_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.sample_n_het_spec.html">sgkit.variables.sample_n_het_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.sample_n_hom_alt_spec.html">sgkit.variables.sample_n_hom_alt_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.sample_n_hom_ref_spec.html">sgkit.variables.sample_n_hom_ref_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.sample_n_non_ref_spec.html">sgkit.variables.sample_n_non_ref_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.sample_pca_component_spec.html">sgkit.variables.sample_pca_component_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.sample_pca_explained_variance_spec.html">sgkit.variables.sample_pca_explained_variance_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.sample_pca_explained_variance_ratio_spec.html">sgkit.variables.sample_pca_explained_variance_ratio_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.sample_pca_loading_spec.html">sgkit.variables.sample_pca_loading_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.sample_pca_projection_spec.html">sgkit.variables.sample_pca_projection_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.sample_ploidy_spec.html">sgkit.variables.sample_ploidy_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_divergence_spec.html">sgkit.variables.stat_divergence_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_diversity_spec.html">sgkit.variables.stat_diversity_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_Fst_spec.html">sgkit.variables.stat_Fst_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_Garud_h1_spec.html">sgkit.variables.stat_Garud_h1_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_Garud_h12_spec.html">sgkit.variables.stat_Garud_h12_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_Garud_h123_spec.html">sgkit.variables.stat_Garud_h123_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_Garud_h2_h1_spec.html">sgkit.variables.stat_Garud_h2_h1_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_genomic_kinship_spec.html">sgkit.variables.stat_genomic_kinship_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_genomic_relationship_spec.html">sgkit.variables.stat_genomic_relationship_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_Hamilton_Kerr_lambda_spec.html">sgkit.variables.stat_Hamilton_Kerr_lambda_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_Hamilton_Kerr_tau_spec.html">sgkit.variables.stat_Hamilton_Kerr_tau_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_hybrid_relationship_spec.html">sgkit.variables.stat_hybrid_relationship_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_hybrid_inverse_relationship_spec.html">sgkit.variables.stat_hybrid_inverse_relationship_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_identity_by_state_spec.html">sgkit.variables.stat_identity_by_state_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_inverse_relationship_spec.html">sgkit.variables.stat_inverse_relationship_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_observed_heterozygosity_spec.html">sgkit.variables.stat_observed_heterozygosity_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_pbs_spec.html">sgkit.variables.stat_pbs_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_pedigree_contribution_spec.html">sgkit.variables.stat_pedigree_contribution_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_pedigree_inbreeding_spec.html">sgkit.variables.stat_pedigree_inbreeding_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_pedigree_inverse_kinship_spec.html">sgkit.variables.stat_pedigree_inverse_kinship_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_pedigree_inverse_relationship_spec.html">sgkit.variables.stat_pedigree_inverse_relationship_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_pedigree_kinship_spec.html">sgkit.variables.stat_pedigree_kinship_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_pedigree_relationship_spec.html">sgkit.variables.stat_pedigree_relationship_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_Tajimas_D_spec.html">sgkit.variables.stat_Tajimas_D_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.stat_Weir_Goudet_beta_spec.html">sgkit.variables.stat_Weir_Goudet_beta_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.traits_spec.html">sgkit.variables.traits_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.variant_allele_spec.html">sgkit.variables.variant_allele_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.variant_allele_count_spec.html">sgkit.variables.variant_allele_count_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.variant_allele_frequency_spec.html">sgkit.variables.variant_allele_frequency_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.variant_allele_total_spec.html">sgkit.variables.variant_allele_total_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.variant_genotype_count_spec.html">sgkit.variables.variant_genotype_count_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.variant_linreg_beta_spec.html">sgkit.variables.variant_linreg_beta_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.variant_call_rate_spec.html">sgkit.variables.variant_call_rate_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.variant_contig_spec.html">sgkit.variables.variant_contig_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.variant_hwe_p_value_spec.html">sgkit.variables.variant_hwe_p_value_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.variant_id_spec.html">sgkit.variables.variant_id_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.variant_n_called_spec.html">sgkit.variables.variant_n_called_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.variant_n_het_spec.html">sgkit.variables.variant_n_het_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.variant_n_hom_alt_spec.html">sgkit.variables.variant_n_hom_alt_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.variant_n_hom_ref_spec.html">sgkit.variables.variant_n_hom_ref_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.variant_n_non_ref_spec.html">sgkit.variables.variant_n_non_ref_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.variant_linreg_p_value_spec.html">sgkit.variables.variant_linreg_p_value_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.variant_ploidy_spec.html">sgkit.variables.variant_ploidy_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.variant_position_spec.html">sgkit.variables.variant_position_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.variant_score_spec.html">sgkit.variables.variant_score_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.variant_linreg_t_value_spec.html">sgkit.variables.variant_linreg_t_value_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.window_contig_spec.html">sgkit.variables.window_contig_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.window_start_spec.html">sgkit.variables.window_start_spec</a></li>
<li class="toctree-l1"><a class="reference internal" href="sgkit.variables.window_stop_spec.html">sgkit.variables.window_stop_spec</a></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../api.html" class="nav-link">API reference</a></li>
    
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">sgkit.simulate_genotype_call_dataset</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                   <section id="sgkit-simulate-genotype-call-dataset">
<h1>sgkit.simulate_genotype_call_dataset<a class="headerlink" href="#sgkit-simulate-genotype-call-dataset" title="Permalink to this heading">#</a></h1>
<dl class="py function">
<dt class="sig sig-object py" id="sgkit.simulate_genotype_call_dataset">
<span class="sig-prename descclassname"><span class="pre">sgkit.</span></span><span class="sig-name descname"><span class="pre">simulate_genotype_call_dataset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_variant</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_sample</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_ploidy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_allele</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_contig</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">missing_pct</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phased</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">additional_variant_fields</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sgkit.simulate_genotype_call_dataset" title="Permalink to this definition">#</a></dt>
<dd><p>Simulate genotype calls and variant/sample data.</p>
<p>Note that the data simulated by this function has no
biological interpretation and that summary statistics
or other methods applied to it will produce meaningless
results.  This function is primarily a convenience on
generating <a class="reference external" href="https://docs.xarray.dev/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v2025.3.2.dev0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">xarray.Dataset</span></code></a> containers so quantities of interest
should be overwritten, where appropriate, within the
context of a more specific application.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>n_variant</strong> <span class="classifier"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span></dt></dt><dd><p>Number of variants to simulate</p>
</dd>
<dt><strong>n_sample</strong> <span class="classifier"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></span></dt></dt><dd><p>Number of samples to simulate</p>
</dd>
<dt><strong>n_ploidy</strong> <span class="classifier"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a> (default: <code class="docutils literal notranslate"><span class="pre">2</span></code>)</span></dt></dt><dd><p>Number of chromosome copies in each sample</p>
</dd>
<dt><strong>n_allele</strong> <span class="classifier"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a> (default: <code class="docutils literal notranslate"><span class="pre">2</span></code>)</span></dt></dt><dd><p>Number of alleles to simulate</p>
</dd>
<dt><strong>n_contig</strong> <span class="classifier"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a> (default: <code class="docutils literal notranslate"><span class="pre">1</span></code>)</span></dt></dt><dd><p>optional
Number of contigs to partition variants with,
controlling values in <code class="docutils literal notranslate"><span class="pre">variant_contig</span></code>. Values
will all be 0 by default when <code class="docutils literal notranslate"><span class="pre">n_contig</span></code> is 1.</p>
</dd>
<dt><strong>seed</strong> <span class="classifier"><span class="annotation terse"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a> | <a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.13)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a></span><span class="annotation full"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.13)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]</span> (default: <code class="docutils literal notranslate"><span class="pre">0</span></code>)</span></dt></dt><dd><p>Seed for random number generation, optional</p>
</dd>
<dt><strong>missing_pct</strong> <span class="classifier"><span class="annotation terse"><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a> | <a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.13)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a></span><span class="annotation full"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.13)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>]</span> (default: <code class="docutils literal notranslate"><span class="pre">None</span></code>)</span></dt></dt><dd><p>The percentage of missing calls, must be within [0.0, 1.0], optional</p>
</dd>
<dt><strong>phased</strong> <span class="classifier"><span class="annotation terse"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a> | <a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.13)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a></span><span class="annotation full"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.13)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>]</span> (default: <code class="docutils literal notranslate"><span class="pre">None</span></code>)</span></dt></dt><dd><p>Whether genotypes are phased, default is unphased, optional</p>
</dd>
<dt><strong>additional_variant_fields</strong> <span class="classifier"><span class="annotation terse"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a> | <a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.13)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a></span><span class="annotation full"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.13)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>]</span> (default: <code class="docutils literal notranslate"><span class="pre">None</span></code>)</span></dt></dt><dd><p>Additional variant fields to add to the dataset as a dictionary of
{field_name: field_dtype}, optional</p>
</dd>
</dl>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.xarray.dev/en/stable/generated/xarray.Dataset.html#xarray.Dataset" title="(in xarray v2025.3.2.dev0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dataset</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>:
A dataset containing the following variables:</p>
<ul class="simple">
<li><p><a class="reference internal" href="sgkit.variables.variant_contig_spec.html#sgkit.variables.variant_contig_spec" title="sgkit.variables.variant_contig_spec"><code class="xref py py-data docutils literal notranslate"><span class="pre">sgkit.variables.variant_contig_spec</span></code></a> (variants)</p></li>
<li><p><a class="reference internal" href="sgkit.variables.variant_position_spec.html#sgkit.variables.variant_position_spec" title="sgkit.variables.variant_position_spec"><code class="xref py py-data docutils literal notranslate"><span class="pre">sgkit.variables.variant_position_spec</span></code></a> (variants)</p></li>
<li><p><a class="reference internal" href="sgkit.variables.variant_allele_spec.html#sgkit.variables.variant_allele_spec" title="sgkit.variables.variant_allele_spec"><code class="xref py py-data docutils literal notranslate"><span class="pre">sgkit.variables.variant_allele_spec</span></code></a> (variants)</p></li>
<li><p><a class="reference internal" href="sgkit.variables.sample_id_spec.html#sgkit.variables.sample_id_spec" title="sgkit.variables.sample_id_spec"><code class="xref py py-data docutils literal notranslate"><span class="pre">sgkit.variables.sample_id_spec</span></code></a> (samples)</p></li>
<li><p><a class="reference internal" href="sgkit.variables.call_genotype_spec.html#sgkit.variables.call_genotype_spec" title="sgkit.variables.call_genotype_spec"><code class="xref py py-data docutils literal notranslate"><span class="pre">sgkit.variables.call_genotype_spec</span></code></a> (variants, samples, ploidy)</p></li>
<li><p><a class="reference internal" href="sgkit.variables.call_genotype_mask_spec.html#sgkit.variables.call_genotype_mask_spec" title="sgkit.variables.call_genotype_mask_spec"><code class="xref py py-data docutils literal notranslate"><span class="pre">sgkit.variables.call_genotype_mask_spec</span></code></a> (variants, samples, ploidy)</p></li>
<li><p><a class="reference internal" href="sgkit.variables.call_genotype_phased_spec.html#sgkit.variables.call_genotype_phased_spec" title="sgkit.variables.call_genotype_phased_spec"><code class="xref py py-data docutils literal notranslate"><span class="pre">sgkit.variables.call_genotype_phased_spec</span></code></a> (variants, samples), if <code class="docutils literal notranslate"><span class="pre">phased</span></code> is not None</p></li>
<li><p>Those specified in <code class="docutils literal notranslate"><span class="pre">additional_variant_fields</span></code>, if provided</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</section>

<div class="section ablog__blog_comments">
   
</div>

                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="sgkit.simulate_genedrop.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">sgkit.simulate_genedrop</p>
      </div>
    </a>
    <a class="right-next"
       href="sgkit.window_by_genome.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">sgkit.window_by_genome</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#sgkit.simulate_genotype_call_dataset"><code class="docutils literal notranslate"><span class="pre">simulate_genotype_call_dataset()</span></code></a></li>
</ul>
  </nav></div>

  <div class="sidebar-secondary-item">
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/generated/sgkit.simulate_genotype_call_dataset.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      Â© Copyright 2020, sgkit developers.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 6.2.1.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>